<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Visium HD Analysis – Spatial Transcriptomics Nanocourse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a7517283740141abf39de03d896ec310.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="../site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="../site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="../site_libs/datatables-binding-0.34.0/datatables.js"></script>
<script src="../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="../site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="../site_libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet">
<script src="../site_libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Spatial Transcriptomics Nanocourse</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/"> 
<span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/spatial_nanocourse/"> 
<span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu"> 
<span class="menu-text">Contact us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#ngs-based-spatial-transcriptomics-data-analysis" id="toc-ngs-based-spatial-transcriptomics-data-analysis" class="nav-link" data-scroll-target="#ngs-based-spatial-transcriptomics-data-analysis">NGS-based Spatial Transcriptomics Data Analysis</a>
  <ul class="collapse">
  <li><a href="#mouse-brain-visium-hd" id="toc-mouse-brain-visium-hd" class="nav-link" data-scroll-target="#mouse-brain-visium-hd">Mouse Brain Visium HD</a></li>
  <li><a href="#preprocessing-data-with-spaceranger" id="toc-preprocessing-data-with-spaceranger" class="nav-link" data-scroll-target="#preprocessing-data-with-spaceranger">Preprocessing Data with Spaceranger</a></li>
  <li><a href="#analysis-workflow" id="toc-analysis-workflow" class="nav-link" data-scroll-target="#analysis-workflow">Analysis workflow</a>
  <ul class="collapse">
  <li><a href="#setting-up" id="toc-setting-up" class="nav-link" data-scroll-target="#setting-up">Setting up</a></li>
  </ul></li>
  <li><a href="#loading-libraries" id="toc-loading-libraries" class="nav-link" data-scroll-target="#loading-libraries">Loading Libraries</a></li>
  <li><a href="#creating-the-seurat-object" id="toc-creating-the-seurat-object" class="nav-link" data-scroll-target="#creating-the-seurat-object">Creating the Seurat Object</a>
  <ul class="collapse">
  <li><a href="#explore-the-object" id="toc-explore-the-object" class="nav-link" data-scroll-target="#explore-the-object">Explore the object</a></li>
  </ul></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality Control</a>
  <ul class="collapse">
  <li><a href="#pre-filtering" id="toc-pre-filtering" class="nav-link" data-scroll-target="#pre-filtering">Pre-filtering</a></li>
  <li><a href="#post-filtering" id="toc-post-filtering" class="nav-link" data-scroll-target="#post-filtering">Post-Filtering</a></li>
  <li><a href="#visualizing-counts-data" id="toc-visualizing-counts-data" class="nav-link" data-scroll-target="#visualizing-counts-data">Visualizing Counts Data</a></li>
  </ul></li>
  <li><a href="#normalize-data" id="toc-normalize-data" class="nav-link" data-scroll-target="#normalize-data">Normalize Data</a></li>
  <li><a href="#unsupervised-clustering" id="toc-unsupervised-clustering" class="nav-link" data-scroll-target="#unsupervised-clustering">Unsupervised Clustering</a></li>
  <li><a href="#project-cluster-labels-back-to-the-full-dataset" id="toc-project-cluster-labels-back-to-the-full-dataset" class="nav-link" data-scroll-target="#project-cluster-labels-back-to-the-full-dataset">Project cluster labels back to the full dataset</a>
  <ul class="collapse">
  <li><a href="#visualizing-the-projected-clusters-on-umap" id="toc-visualizing-the-projected-clusters-on-umap" class="nav-link" data-scroll-target="#visualizing-the-projected-clusters-on-umap">Visualizing the projected clusters on UMAP</a></li>
  <li><a href="#visualizing-projected-clusters-on-the-image" id="toc-visualizing-projected-clusters-on-the-image" class="nav-link" data-scroll-target="#visualizing-projected-clusters-on-the-image">Visualizing projected clusters on the image</a></li>
  </ul></li>
  <li><a href="#spatially-informed-clustering" id="toc-spatially-informed-clustering" class="nav-link" data-scroll-target="#spatially-informed-clustering">Spatially-informed Clustering</a></li>
  <li><a href="#cell-type-annotation" id="toc-cell-type-annotation" class="nav-link" data-scroll-target="#cell-type-annotation">Cell Type Annotation</a>
  <ul class="collapse">
  <li><a href="#sketch-and-process-the-spatial-query-dataset" id="toc-sketch-and-process-the-spatial-query-dataset" class="nav-link" data-scroll-target="#sketch-and-process-the-spatial-query-dataset">1) Sketch and process the spatial query dataset</a></li>
  <li><a href="#load-and-format-the-reference-dataset" id="toc-load-and-format-the-reference-dataset" class="nav-link" data-scroll-target="#load-and-format-the-reference-dataset">2) Load and format the reference dataset</a></li>
  <li><a href="#apply-rctd-to-deconvolute-the-sketched-cortical-cells-and-annotate-them" id="toc-apply-rctd-to-deconvolute-the-sketched-cortical-cells-and-annotate-them" class="nav-link" data-scroll-target="#apply-rctd-to-deconvolute-the-sketched-cortical-cells-and-annotate-them">3) Apply RCTD to deconvolute the “sketched” cortical cells and annotate them</a></li>
  <li><a href="#project-rctd-labels-onto-all-cortical-cells" id="toc-project-rctd-labels-onto-all-cortical-cells" class="nav-link" data-scroll-target="#project-rctd-labels-onto-all-cortical-cells">4) Project RCTD labels onto all cortical cells</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#save-r-session" id="toc-save-r-session" class="nav-link" data-scroll-target="#save-r-session">Save R Session</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Visium HD Analysis</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Alex Bartlett, Meeta Mistry, and Will Gammerdinger </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>Contributors: Alex Bartlett, Meeta Mistry and Will Gammerdinger</p>
<p>Approximate time: 2 hours and 45 minutes</p>
<section id="learning-objectives" class="level1">
<h1>Learning Objectives</h1>
<p>In this lesson, we will:</p>
<ul>
<li>Describe the elements of the Seurat object and how they are generated</li>
<li>Visually inspect and compare spatial scRNA-seq data before and after filtering</li>
<li>Execute clustering workflows and visualize results on a tissue section</li>
</ul>
</section>
<section id="ngs-based-spatial-transcriptomics-data-analysis" class="level1">
<h1>NGS-based Spatial Transcriptomics Data Analysis</h1>
<section id="mouse-brain-visium-hd" class="level2">
<h2 class="anchored" data-anchor-id="mouse-brain-visium-hd">Mouse Brain Visium HD</h2>
<p>The Visium HD platform is compatible with human and mouse fresh frozen, fixed frozen, and formalin-fixed parrafin-embedded (FFPE) tissue sections. For this lesson, we will be working with data from a fresh frozen coronal section of a mouse brain sample.</p>
<p>Each Visium HD slide has the same 6.5 x 6.5mm capture area as previous Visium products but is covered with about 11 million tiles. These 2µm x 2µm squares are arrayed in a continuous lawn across the entire capture area. The squares are each uniquely barcoded with an oligonucleotide and contain probes allowing for the detection of the full coding transcriptome.</p>
<p align="center">
<img src="../img/visium_hd_slide_graphic.png" width="800">
</p>
</section>
<section id="preprocessing-data-with-spaceranger" class="level2">
<h2 class="anchored" data-anchor-id="preprocessing-data-with-spaceranger">Preprocessing Data with Spaceranger</h2>
<p>Sequencing facilities often output scRNA-seq data, including spatial scRNA-seq data, in FASTQ format. Because this is Visium HD data from 10X Genomics, we used their proprietary preprocessing software <a href="https://www.10xgenomics.com/support/software/space-ranger/latest">Space Ranger</a> to process the FASTQ files into a count matrix and other images. Specifically, the <code>spaceranger count</code> command aligns the reads in the FASTQ files against a transcriptomic reference and provides their spatial location using the oligonucleotide barcode.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click here to see an example of how to run <code>spaceranger count</code>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>A sample command for running <code>spaceranger count</code> is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">spaceranger</span> count <span class="at">--id</span><span class="op">=</span>hd_count <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">--transcriptome</span><span class="op">=</span>/path/to/refdata-gex-GRCh38-2020-A <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">--fastqs</span><span class="op">=</span>/path/to/fastq <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">--probe-set</span><span class="op">=</span>/path/to/Visium_Human_Transcriptome_Probe_Set_v2.0_GRCh38-2020-A.csv <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">--slide</span><span class="op">=</span>H1-YD7CDZK <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">--area</span><span class="op">=</span>A1 <span class="dt">\</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>   <span class="at">--cytaimage</span><span class="op">=</span>/path/to/CAVG10539_2023-11-16_14-56-24_APPS115_H1-YD7CDZK_A1_S11088.tif <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>   <span class="at">--image</span><span class="op">=</span>/path/to/APPS115_11088_rescan_01.btf <span class="dt">\</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>   <span class="at">--create-bam</span><span class="op">=</span>false</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>Note that Space Ranger requires a Linux system with at least 32 cores, 64GB of RAM, and 1TB of disk space.</p>
<p>When <code>spaceranger count</code> completes successfully, it will generate output similar to the following, which will enable the analyst to perform further analysis in R or Python or using the proprietary Loupe browser from 10X Genomics.</p>
<p align="center">
<img src="../img/spaceranger_output.svg" alt="spaceranger output" width="600">
</p>
<p>In the Visium HD assay, in addition to providing data at the level of the 2µm x 2µm squares, Space Ranger also bins the 2µm x 2µm squares into 8µm x 8µm and 16µm x 16µm bins. Most of the above output is produced for each binning resolution.</p>
<p>While the single-digit micron resolution is a big technological improvement over original Visium’s original ∼55μm spots, the higher resolution also presents challenges. Having access to 2μm bins along with matching morphology information provides a great opportunity for reconstructing single cells from the data, which is undoubtedly very powerful. However, because the 2µm x 2µm squares (and even the 8µm x 8µm bins) are so small, there is a potential for very little biological signal to be captured per bin. Additionally, the sheer number of bins at these higher resolutions can present challenges in terms of computational time and resources. <strong>For the purposes of this lesson, we will use the 16µm x 16µm bins.</strong></p>
<p>We can view and explore the web summary HTML of our data found in the “reports” folder of your project.</p>
</section>
<section id="analysis-workflow" class="level2">
<h2 class="anchored" data-anchor-id="analysis-workflow">Analysis workflow</h2>
<p align="center">
<img src="../img/Full_workflow.png" width="800">
</p>
<section id="setting-up" class="level3">
<h3 class="anchored" data-anchor-id="setting-up">Setting up</h3>
<p>For this module, we will be working within an RStudio project. In order to follow along you should have <strong>downloaded the R project</strong>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you haven’t done this already, the project is located in “Dataset for workshop” -&gt; “Day 2- NGS-based- VisiumHD” in the course DropBox.</p>
</div>
</div>
<p>Once downloaded, you should see a file called <code>visiumHD_nanocourse.zip</code> on your computer (likely, in your <code>Downloads</code> folder).</p>
<ol type="1">
<li><p><strong>Unzip this file.</strong> This will result in a folder of the same name.</p></li>
<li><p><strong>Move the folder to the location on your computer where you would like to perform the analysis.</strong></p></li>
<li><p><strong>Open up the folder.</strong> The contents will look like the screenshot below:</p>
<p align="center">
</p><p><img src="../img/proj_screenshot.png" width="500"></p>
<p></p></li>
<li><p><strong>Locate the <code>.Rproj file</code> and double-click on it.</strong> This will open up RStudio with the “visiumHD_nanocourse” project loaded.</p></li>
<li><p><strong>Open a new Rscript file.</strong></p></li>
<li><p><strong>Start with some comments to indicate what this file is going to contain:</strong></p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># February 26th, 2025</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatial Transcriptomics: Session 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="7" type="1">
<li><p><strong>Save the Rscript in the <code>code</code> folder as <code>visiumHD.R</code>.</strong> Your working directory should look something like this:</p>
<p align="center">
</p><p><img src="../img/Code_folder.gif" width="700"></p>
<p></p></li>
</ol>
</section>
</section>
<section id="loading-libraries" class="level2">
<h2 class="anchored" data-anchor-id="loading-libraries">Loading Libraries</h2>
<p>Next, we will need to be sure to load the libraries that we will be using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Banksy)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(quadprog)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spacexr)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Increases the size of the default vector</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize=</span> <span class="dv">2000000000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="creating-the-seurat-object" class="level2">
<h2 class="anchored" data-anchor-id="creating-the-seurat-object">Creating the Seurat Object</h2>
<p>The Seurat object is a custom list-like object that has well-defined spaces to store specific information/data for single-cell experiments, including spatial experiments and Visium HD.</p>
<p>The Seurat package provides a function <code>Load10X_Spatial()</code> to easily create a Seurat object from the output of Space Ranger. The <code>Load10X_Spatial</code> function takes as input the feature matrix and the low-resolution tissue image from the output of Space Ranger and generates a Seurat object containing both gene-level counts and spatial information.</p>
<p><strong>We will not have you run this code</strong>, as this can take some time and the Space Ranger output files are quite large to share. Instead, you will load the pre-made Seurat object.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click here to see the R code used to create the Seurat object
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>localdir <span class="ot">&lt;-</span> <span class="st">'../path/to/spaceranger/outs/'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Load the raw feature matrix</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>object <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(<span class="at">data.dir =</span> localdir,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">filename =</span> <span class="st">'raw_feature_bc_matrix.h5'</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">bin.size =</span> <span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<section id="explore-the-object" class="level3">
<h3 class="anchored" data-anchor-id="explore-the-object">Explore the object</h3>
<p>Let’s read in the Seurat object and talk about some very basic slots that we will be accessing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in Seurat object</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>object <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">'data_processed/MsBrain_FF-A1_subset_misc.qs'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can print the Seurat object by using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>object</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32285 features across 43167 samples within 1 assay 
Active assay: Spatial.016um (32285 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>Now we can examine its major features, which we will add to and alter throughout the lesson:</p>
<p align="center">
<img src="../img/Base_seurat_object_labelled.png" width="1000">
</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>There are 3 things about our Seurat object printout that would be different if we were using the 8µm x 8µm binning instead of the 16µm x 16µm binning. What are these three differences?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click here to use an app that lets you explore different bin sizes for this Seurat object
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p align="center">
<iframe src="https://hcbc.connect.hms.harvard.edu/Spatial_resolution_question_1/?showcase=0" width="600px" height="250px" data-external="1">
</iframe>
</p>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="quality-control" class="level2">
<h2 class="anchored" data-anchor-id="quality-control">Quality Control</h2>
<p>The main objective of quality control is to filter the data so that we include only data from bins that are of high-quality. This makes it so that when we cluster our bins, it is easier to identify distinct cell type populations.</p>
<p align="center">
<img src="../img/QC_workflow.png" width="800">
</p>
<p>In Visium HD data, the main challenge is in <strong>delineating bins that are poor quality from bins containing reads from less complex cells</strong>. If you expect a particular cell type in your dataset to be less transcriptionally active as compared other cell types in your dataset, the bins underneath this cell type will naturally have fewer detected genes and transcripts. However, having fewer detected genes and transcripts can also be a technical artifact and not a result of biological signal.</p>
<p>Various metrics can be used to filter low-quality bins from high-quality ones, including:</p>
<ul>
<li><p><strong>UMI counts per bin</strong> - This is the number of unique transcripts detected per bin. Because the bins are very small, this number is less than what we would expect for non-spatial scRNA-seq data.</p></li>
<li><p><strong>Genes detected per bin</strong> - This is the number of unique genes detected per bin. Again, because the bins are very small, this number is less than what we would expect for non-spatial scRNA-seq data.</p></li>
<li><p><strong>Complexity (novelty score)</strong> - The novelty score is computed as shown below:</p>
<p align="center">
</p><p><img src="https://latex.codecogs.com/svg.image?\text{Complexity&amp;space;Score}=\frac{\text{Number&amp;space;of&amp;space;Genes}}{\text{Number&amp;space;of&amp;space;UMIs}}"></p>
<p></p>
<p>If there are many captured transcripts (high nUMI) and a low number of genes detected in a bin, this likely means that you only captured a low number of genes and simply sequenced transcripts from those lower number of genes over and over again. These low complexity (low novelty) bins could represent a specific cell type (i.e.&nbsp;red blood cells, which lack a typical transcriptome), or could be due to an artifact or contamination. Generally, we expect the complexity score to be above 0.80 for good-quality bins.</p></li>
<li><p><strong>Mitochondrial counts ratio</strong> - This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells. We define poor-quality samples for mitochondrial counts as bins which surpass the 0.2 mitochondrial ratio threshold, unless of course you are expecting this in your sample. This ratio is computed as:</p>
<p align="center">
</p><p><img src="https://latex.codecogs.com/svg.image?\text{Mitochondiral&amp;space;Ratio}=\frac{\text{Number&amp;space;of&amp;space;reads&amp;space;aligning&amp;space;to&amp;space;mitochondrial&amp;space;genes}}{\text{Total&amp;space;reads}}"></p>
<p></p></li>
</ul>
<p>Let’s take a quick look at the data and make a decision on whether we need to apply any filtering. We will examine the distributions of UMI counts per bin and genes detected per bin to determine reasonable thresholds for those metrics to implement during QC filtering.</p>
<section id="pre-filtering" class="level3">
<h3 class="anchored" data-anchor-id="pre-filtering">Pre-filtering</h3>
<p>To create some plots first, we will need to create a metadata object using this command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>object_meta <span class="ot">&lt;-</span> object<span class="sc">@</span>meta.data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can plot the number of UMIs (nUMI) and the number of genes (nGene) side-by-side. For both of the plots, we expect to see a bimodal distribution, with one peak representing bins containing lower-quality cells with fewer genes and UMIs and another peak representing bins containing healthy cells with more genes and UMIs. Ideally, the peak representing lower-quality and dying cells is small and the peak representing healthy cells is large.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot for nUMI</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>dist_counts_before <span class="ot">&lt;-</span> object_meta <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nCount_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of UMIs per bin"</span>) <span class="sc">+</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Pre-QC UMIs/Bin'</span>) <span class="sc">+</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot for nGene</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>dist_features_before <span class="ot">&lt;-</span> object_meta <span class="sc">%&gt;%</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nFeature_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of genes per bin"</span>) <span class="sc">+</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Pre-QC Genes/Bin'</span>) <span class="sc">+</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>dists_before <span class="ot">&lt;-</span> dist_counts_before <span class="sc">|</span> dist_features_before</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>dists_before</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using the distribution plots in the app below, what do you think would be good minimum thresholds for nGene and nUMI?</p>
<p align="center">
<iframe src="https://hcbc.connect.hms.harvard.edu/Spatial_threshold_question_2/?showcase=0" width="800px" height="410px" data-external="1">
</iframe>
</p>
</div>
</div>
</section>
<section id="post-filtering" class="level3">
<h3 class="anchored" data-anchor-id="post-filtering">Post-Filtering</h3>
<p>We will apply very minimal filtering here, with nUMI &gt; 100 and nGene &gt; 100. It has been shown that low expression can be biologically meaningful for spatial context so we won’t be as stringent as we normally are with scRNA-seq.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a filtered object</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(object, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">subset =</span> (nCount_Spatial<span class="fl">.016</span>um <span class="sc">&gt;</span> <span class="dv">100</span>) <span class="sc">&amp;</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                        (nFeature_Spatial<span class="fl">.016</span>um <span class="sc">&gt;</span> <span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can create similar plots with filtered data. As expected, we see that the small left peak in the distribution has vanished, leaving the higher quality bins, which are the majority of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new metadata data frame </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>object_filt_meta <span class="ot">&lt;-</span> object_filt<span class="sc">@</span>meta.data</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot nUMI</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>dist_counts_after <span class="ot">&lt;-</span> object_filt_meta <span class="sc">%&gt;%</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nCount_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of UMIs per bin"</span>) <span class="sc">+</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'PostQC UMIs/Bin'</span>) <span class="sc">+</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot nGene</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>dist_features_after <span class="ot">&lt;-</span> object_filt_meta <span class="sc">%&gt;%</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nFeature_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of genes per bin"</span>) <span class="sc">+</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'PostQC Genes/Bin'</span>) <span class="sc">+</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine plots side-by-side</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>dists_after <span class="ot">&lt;-</span> dist_counts_after <span class="sc">|</span> dist_features_after</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>dists_after</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualizing-counts-data" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-counts-data">Visualizing Counts Data</h3>
<p>We can visualize the number of UMIs and gene counts per bin, both as a distribution and layered on top of the tissue image. Let’s start with a violin plot to look at the distribution of UMI counts and gene counts. The input is our post-filtered dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Violin plot of UMI counts</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>vln_counts_after <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(object_filt, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">features =</span> <span class="st">"nCount_Spatial.016um"</span>, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">pt.size =</span> <span class="dv">0</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">group.by =</span> <span class="st">'orig.ident'</span>) <span class="sc">+</span> </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'nUMI'</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">''</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">15000</span>))</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Violin plot of gene counts</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>vln_features_after <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(object_filt, </span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>                            <span class="at">features =</span> <span class="st">"nFeature_Spatial.016um"</span>, </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>                            <span class="at">pt.size =</span> <span class="dv">0</span>, </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>                            <span class="at">group.by =</span> <span class="st">'orig.ident'</span>) <span class="sc">+</span> </span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'nGene'</span>) <span class="sc">+</span>  <span class="fu">xlab</span>(<span class="st">''</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">15000</span>))</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot both side by side</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>vln_counts_after <span class="sc">|</span> vln_features_after</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We see that both distributions have a similar peak but the nUMI distribution has a much longer tail. This is expected, because while the small physical size of the bins means that most genes will be detected only once or twice, a minority of bins under very transcriptionally active cells may exhibit multiple transcripts of the same gene.</p>
<p>Next, we can look at the same metrics and the distribution on the actual image itself. Note that many spots have very few counts, in part due to low cellular density or cell types with low complexity in certain tissue regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing UMI count across the image</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>image_counts <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(object_filt, </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">feature =</span> <span class="st">'nCount_Spatial.016um'</span>, </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">pt.size.factor =</span> <span class="dv">8</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing gene count across the image</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>image_features <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(object_filt, </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">features =</span> <span class="st">"nFeature_Spatial.016um"</span>, </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">pt.size.factor =</span> <span class="dv">8</span>) </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the two side-by-side</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>image_counts <span class="sc">|</span> image_features</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="normalize-data" class="level2">
<h2 class="anchored" data-anchor-id="normalize-data">Normalize Data</h2>
<p>Normalization is important in order to make expression counts comparable across genes and/or samples. We note that the best normalization methods for spatial data are still being developed and evaluated. In particular, <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03241-7">Bhuva et. al</a> tested a variety of normalization techniques and found that normalizing by the number of transcripts detected per bin negatively affects spatial domain identification because total detections per bin can represent real biology. We are cognizant of this, but as discussed earlier, it can be challenging to determine whether a bin has few detections because of a technical artifact or biological signal. In the absence of normalization, this lack of signal will strongly affect clustering regardless of whether it is technical or biological. For this reason, we apply a standard log-transformed library size normalization to our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(object_filt, <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After normalization, we can call our Seurat object with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32285 features across 41818 samples within 1 assay 
Active assay: Spatial.016um (32285 features, 0 variable features)
 2 layers present: counts, data
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>And we can see that there is now a new “data” layer in the Seurat object.</p>
<p align="center">
<img src="../img/Seurat_object_with_normalized_data_labelled.png" width="700">
</p>
</section>
<section id="unsupervised-clustering" class="level2">
<h2 class="anchored" data-anchor-id="unsupervised-clustering">Unsupervised Clustering</h2>
<p>The authors of the Seurat package recommend the Seurat v5 sketch clustering workflow because it exhibits improved performance for large datasets, especially for identifying rare and spatially-restricted groups. Sketch-based analyses aim to “subsample” large datasets in a way that preserves rare populations.</p>
<p align="center">
<img src="../img/Standard_clustering.png" width="800">
</p>
<p>We will start by defining a set of highly variable genes. <em>Note that this is being done on all bins in our object.</em> Using this list of genes will help us to quantify the variability and similarity between bins.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(object_filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can examine our Seurat object and see that <code>FindVariableFeatures()</code> has added 2,000 variable features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32285 features across 41818 samples within 1 assay 
Active assay: Spatial.016um (32285 features, 2000 variable features)
 2 layers present: counts, data
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p align="center">
<img src="../img/Seurat_object_variable_features_labelled.png" width="700">
</p>
<p>Next, we select 10,000 cells and create a new sub-sampled “sketch” assay using the <code>SketchData()</code> function. The function takes a normalized single-cell dataset containing a set of variable features and returns a Seurat object with a new assay (sketch), consisting of 10,000 bins selected based off a “leverage score” for each bin. The leverage score reflects the magnitude of its contribution to the gene-covariance matrix, and its importance to the overall dataset, with rare populations earning a higher leverage score. This means that our 10,000 cells selected for the sketch will oversample rare populations, retaining the biological complexity of the sample, while drastically compressing the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we select 10,000 cells and create a new 'sketch' assay</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">SketchData</span>(</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> object_filt,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncells =</span> <span class="dv">10000</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"LeverageScore"</span>,</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have the sketched data, we can call the Seurat object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
64570 features across 41818 samples within 2 assays 
Active assay: sketch (32285 features, 2000 variable features)
 2 layers present: counts, data
 1 other assay present: Spatial.016um
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>We will see that there are four major changes that have taken place:</p>
<ul>
<li>The number of features in the second line has double, because we have added a new assay</li>
<li>Accordingly, the number of assays has increased from one to two</li>
<li>The active assay has change from <code>Spatial.016um</code> to <code>sketch</code></li>
<li>There is a new line listing additional assays that exist in the Seurat object</li>
</ul>
<p align="center">
<img src="../img/Seurat_object_sketch_labelled.png" width="700">
</p>
<p>We can also see that the leverage score has been added as a column to the metadata of our object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(object_filt<span class="sc">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-8213c8232d31420bf992" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8213c8232d31420bf992">{"x":{"filter":"none","vertical":false,"data":[["s_016um_00010_00130-1","s_016um_00010_00184-1","s_016um_00010_00196-1","s_016um_00010_00197-1","s_016um_00010_00198-1","s_016um_00010_00205-1","s_016um_00010_00208-1","s_016um_00010_00217-1","s_016um_00011_00074-1","s_016um_00011_00075-1","s_016um_00011_00076-1","s_016um_00011_00077-1","s_016um_00011_00097-1","s_016um_00011_00098-1","s_016um_00011_00119-1","s_016um_00011_00130-1","s_016um_00011_00135-1","s_016um_00011_00164-1","s_016um_00011_00165-1","s_016um_00011_00166-1","s_016um_00011_00167-1","s_016um_00011_00168-1","s_016um_00011_00169-1","s_016um_00011_00170-1","s_016um_00011_00171-1","s_016um_00011_00172-1","s_016um_00011_00173-1","s_016um_00011_00174-1","s_016um_00011_00175-1","s_016um_00011_00176-1"],["s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s"],[113,110,113,116,136,125,105,112,133,209,210,136,132,128,110,113,143,154,109,110,137,112,132,117,143,131,131,149,233,162],[110,109,112,113,131,120,101,103,128,200,205,129,119,109,108,108,135,147,102,109,135,109,132,117,141,121,125,145,222,149],[1.014817726388032,0.8205218946758598,0.8697821714942078,0.7025807323369158,0.9083597228005351,1.113871348620476,1.005015429894662,1.531798796455271,1.241524123624483,1.434526762848253,1.226601388785213,0.5794057900350176,2.901698137547762,2.007511613487927,0.5359644799724732,2.704265661958694,1.059081859686901,1.081797860857503,1.587927387405159,0.7951949874305146,1.094563187107874,0.9136259126205523,0.6502673348171317,0.8852512091251488,0.7039981022761093,1.983810652331956,0.8063526907498312,1.251575432322209,1.021190625203122,1.299358348842617]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>orig.ident<\/th>\n      <th>nCount_Spatial.016um<\/th>\n      <th>nFeature_Spatial.016um<\/th>\n      <th>leverage.score<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"orig.ident","targets":1},{"name":"nCount_Spatial.016um","targets":2},{"name":"nFeature_Spatial.016um","targets":3},{"name":"leverage.score","targets":4}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>Next, we will peform a standard clustering workflow on our sketch of 10,000 cells:</p>
<ul>
<li><code>FindVariableFeatures</code>: As before, this generates a list of highly variable genes, which may be slighly different for the sketched dataset than for the full dataset</li>
<li><code>ScaleData</code>: Highly variable genes will be confounded with the most highly expressed genes, so we need to adjust for this</li>
<li><code>RunPCA</code>: Perform a principal component analysis using our scaled data and variable genes. This will emphasize variation in gene expression as well as similarity across bins</li>
<li><code>FindNeighbors</code>: Determine the Euclidean distance between bins in PCA space</li>
<li><code>FindClusters</code>: Iteratively group bins together based on neighborhood distances. Higher resolution will yield more groups.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(object_filt)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(object_filt)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(object_filt, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction.name =</span> <span class="st">"pca.sketch"</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(object_filt, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction =</span> <span class="st">"pca.sketch"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(object_filt, </span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                            <span class="at">cluster.name =</span> <span class="st">"seurat_cluster.sketched"</span>, </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">resolution =</span> .<span class="dv">65</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 10000
Number of edges: 401729

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9013
Number of communities: 20
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p>Finally, let’s create a UMAP using the principal components as input. UMAP is a method that aims to place cells with similar local neighborhoods in high-dimensional space together in low-dimensional space, which is useful for visualizing our newly calculated clusters. We observe good separation between groups annotated as separate clusters, which is sign that our clustering indeed represents various cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(object_filt, <span class="at">reduction =</span> <span class="st">"pca.sketch"</span>, </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">reduction.name =</span> <span class="st">"umap.sketch"</span>, <span class="at">return.model =</span> T, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(object_filt, <span class="at">reduction =</span> <span class="st">"umap.sketch"</span>, <span class="at">label =</span> T, </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">cols =</span> <span class="st">'polychrome'</span>) <span class="sc">+</span> </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Sketched clustering"</span>) <span class="sc">+</span> </span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also examine our object after these manipulations and note the addition of the scale.data layer as well as the sketch PCA and UMAP dimensional reductions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
64570 features across 41818 samples within 2 assays 
Active assay: sketch (32285 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial.016um
 2 dimensional reductions calculated: pca.sketch, umap.sketch
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>Should return:</p>
<p align="center">
<img src="../img/Seurat_object_scale_data_PCA_UMAP_labelled.png" width="700">
</p>
</section>
<section id="project-cluster-labels-back-to-the-full-dataset" class="level2">
<h2 class="anchored" data-anchor-id="project-cluster-labels-back-to-the-full-dataset">Project cluster labels back to the full dataset</h2>
<p>Now that we have our clusters and dimensional reductions from our sketched dataset, we need to extend these to the full dataset. The <code>ProjectData</code> function projects all the bins in the dataset (the <code>Spatial.016um</code> assay) onto the <code>sketch</code> assay.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">ProjectData</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> object_filt,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"Spatial.016um"</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.reduction =</span> <span class="st">"full.pca.sketch"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.reduction =</span> <span class="st">"pca.sketch"</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.model =</span> <span class="st">"umap.sketch"</span>,</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">seurat_cluster.projected =</span> <span class="st">"seurat_cluster.sketched"</span>)</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using the sketch PCA and UMAP, the <code>ProjectData</code> function returns a Seurat object that includes: * <strong>Dimensional reduction (PCA)</strong> - The <code>full.pca.sketch</code> dimensional reduction extends the PCA reduction on the sketched cells to all bins in the dataset * <strong>Dimensional reduction (UMAP)</strong> - The <code>full.umap.sketch</code> dimensional reduction extends the UMAP reduction on the sketched cells to all bins in the dataset * <strong>Cluster labels</strong> - The <code>seurat_cluster.projected</code> column in the object metadata now labels all cells in the dataset with one of the cluster labels derived from the sketched cells</p>
<p>We can now see the additional full-dataset reductions in the object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
64570 features across 41818 samples within 2 assays 
Active assay: sketch (32285 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial.016um
 4 dimensional reductions calculated: pca.sketch, umap.sketch, full.pca.sketch, full.umap.sketch
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>Should return:</p>
<p align="center">
<img src="../img/Seurat_object_projected_data_labelled.png" width="700">
</p>
<p>Note that a score for the projection of each bin will be saved as a column in the metadata. Actually opening up the metadata again gives the opportunity to look at the <code>seurat_cluster.sketched</code> column and see many NA values, because it was only calculated for 10,000 bins. The <code>seurat_cluster.projected</code> shows values for every bin.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(object_filt<span class="sc">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-98ea9d8fa7eed1bb98e6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-98ea9d8fa7eed1bb98e6">{"x":{"filter":"none","vertical":false,"data":[["s_016um_00010_00130-1","s_016um_00010_00184-1","s_016um_00010_00196-1","s_016um_00010_00197-1","s_016um_00010_00198-1","s_016um_00010_00205-1","s_016um_00010_00208-1","s_016um_00010_00217-1","s_016um_00011_00074-1","s_016um_00011_00075-1","s_016um_00011_00076-1","s_016um_00011_00077-1","s_016um_00011_00097-1","s_016um_00011_00098-1","s_016um_00011_00119-1","s_016um_00011_00130-1","s_016um_00011_00135-1","s_016um_00011_00164-1","s_016um_00011_00165-1","s_016um_00011_00166-1","s_016um_00011_00167-1","s_016um_00011_00168-1","s_016um_00011_00169-1","s_016um_00011_00170-1","s_016um_00011_00171-1","s_016um_00011_00172-1","s_016um_00011_00173-1","s_016um_00011_00174-1","s_016um_00011_00175-1","s_016um_00011_00176-1"],["s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s","s"],[113,110,113,116,136,125,105,112,133,209,210,136,132,128,110,113,143,154,109,110,137,112,132,117,143,131,131,149,233,162],[110,109,112,113,131,120,101,103,128,200,205,129,119,109,108,108,135,147,102,109,135,109,132,117,141,121,125,145,222,149],[1.014817726388032,0.8205218946758598,0.8697821714942078,0.7025807323369158,0.9083597228005351,1.113871348620476,1.005015429894662,1.531798796455271,1.241524123624483,1.434526762848253,1.226601388785213,0.5794057900350176,2.901698137547762,2.007511613487927,0.5359644799724732,2.704265661958694,1.059081859686901,1.081797860857503,1.587927387405159,0.7951949874305146,1.094563187107874,0.9136259126205523,0.6502673348171317,0.8852512091251488,0.7039981022761093,1.983810652331956,0.8063526907498312,1.251575432322209,1.021190625203122,1.299358348842617],[null,null,null,"12",null,null,null,null,"12",null,null,null,null,null,null,null,null,"12",null,null,null,null,"12",null,null,"12",null,"12",null,"12"],["0","0","12","12","12","12","12","0","0","0","12","0","0","0","0","12","12","12","12","0","12","0","12","12","12","12","0","12","12","12"],[0.7467201156111981,0.6953189441567776,0.520957255636001,0.5386905498086193,0.8289968295423074,0.5927143558197466,0.6725333894882957,0.9400605430577138,0.5447480195031096,0.6104231256592125,0.6994792413673974,0.6022748421638892,0.6904961467514872,0.9664716346999849,0.6052876820321464,0.5406105996860472,0.7049335416147086,0.6632967094509646,0.6938230261160363,0.5387987273928214,0.6832821725075865,0.643596280225364,0.6997898095804197,0.6164056136621444,0.6423525881902905,0.7188979885479108,0.5390039871457251,0.6304374990323328,0.822304335185893,0.7893161605302884]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>orig.ident<\/th>\n      <th>nCount_Spatial.016um<\/th>\n      <th>nFeature_Spatial.016um<\/th>\n      <th>leverage.score<\/th>\n      <th>seurat_cluster.sketched<\/th>\n      <th>seurat_cluster.projected<\/th>\n      <th>seurat_cluster.projected.score<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"orig.ident","targets":1},{"name":"nCount_Spatial.016um","targets":2},{"name":"nFeature_Spatial.016um","targets":3},{"name":"leverage.score","targets":4},{"name":"seurat_cluster.sketched","targets":5},{"name":"seurat_cluster.projected","targets":6},{"name":"seurat_cluster.projected.score","targets":7}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>Should return:</p>
<section id="visualizing-the-projected-clusters-on-umap" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-the-projected-clusters-on-umap">Visualizing the projected clusters on UMAP</h3>
<p>We can now visualize our clusters from the projected assignments. The UMAP plot now contains more points, which is expected because we are now visualizing the full dataset rather than our 10,000 bin sketch. Nonetheless, we can see that the full dataset is still well-representated by the projected dimensional reduction and clustering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># switch to full dataset assay</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(object_filt) <span class="ot">&lt;-</span> <span class="st">"Spatial.016um"</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Change the idents to the projected cluster assignments</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(object_filt) <span class="ot">&lt;-</span> <span class="st">"seurat_cluster.projected"</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(object_filt, <span class="at">reduction =</span> <span class="st">"full.umap.sketch"</span>, <span class="at">label =</span> T, <span class="at">raster =</span> F, </span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">cols =</span> <span class="st">'polychrome'</span>) <span class="sc">+</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Projected clustering"</span>) <span class="sc">+</span> </span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualizing-projected-clusters-on-the-image" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-projected-clusters-on-the-image">Visualizing projected clusters on the image</h3>
<p>In order to see the clusters superimposed on our image we can use the <code>SpatialDimPlot()</code> function. We will also set the color palette and convert the cluster assignments to a factor so they are ordered numerically rather than lexicographically in the figure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Arrange so clusters get listed in numerical order</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>object_filt<span class="sc">$</span>seurat_cluster.projected <span class="ot">&lt;-</span> object_filt<span class="sc">$</span>seurat_cluster.projected <span class="sc">%&gt;%</span> </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  as.numeric <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Set color palette</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected)),</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected))</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(object_filt, </span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">group.by =</span> <span class="st">'seurat_cluster.projected'</span>, </span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">pt.size.factor =</span> <span class="dv">8</span>, <span class="at">cols =</span> color_pal) <span class="sc">+</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">ncol=</span><span class="dv">2</span>))</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="spatially-informed-clustering" class="level2">
<h2 class="anchored" data-anchor-id="spatially-informed-clustering">Spatially-informed Clustering</h2>
<p><a href="https://www.nature.com/articles/s41588-024-01664-3">BANKSY</a> is another method for performing clustering. Unlike Seurat, BANKSY takes into account not only an individual bin’s expression pattern but also the mean and the gradient of gene expression levels in a bin’s broader neighborhood. This makes it valuable for identifying and defining spatial regions of interest.</p>
<p>We use the <code>RunBanksy</code> function to create a new “BANKSY” assay based on a default of the 4,000 most highly variable features, which can be used for dimensionality reduction and clustering. Two parameters of importance are: * <code>k_geom</code> - Number of bins to consider for the local neighborhood. Larger values will yield larger domains. * <code>lambda</code> - Influence of the neighborhood. Larger values yield more spatially coherent domains. The authors recommend using 0.8 to identify broader spatial domains.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run Banksy</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunBanksy</span>(object_filt, <span class="at">lambda =</span> <span class="fl">0.8</span>, <span class="at">verbose =</span> T,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>, <span class="at">slot =</span> <span class="st">'data'</span>, <span class="at">k_geom =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see the new BANKSY assay in our object by calling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>object_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
68570 features across 41818 samples within 3 assays 
Active assay: BANKSY (4000 features, 0 variable features)
 2 layers present: data, scale.data
 2 other assays present: Spatial.016um, sketch
 4 dimensional reductions calculated: pca.sketch, umap.sketch, full.pca.sketch, full.umap.sketch
 1 spatial field of view present: slice1.016um</code></pre>
</div>
</div>
<p>Which should return:</p>
<p align="center">
<img src="../img/Seurat_object_BANKSY_labelled.png" width="700">
</p>
<p>We perform a simplified clustering workflow on the BANKSY assay.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(object_filt, <span class="at">assay =</span> <span class="st">"BANKSY"</span>, </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction.name =</span> <span class="st">"pca.banksy"</span>, </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">features =</span> <span class="fu">rownames</span>(object_filt), <span class="at">npcs =</span> <span class="dv">30</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(object_filt, <span class="at">reduction =</span> <span class="st">"pca.banksy"</span>, </span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(object_filt, <span class="at">cluster.name =</span> <span class="st">"banksy_cluster"</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>                            <span class="at">resolution =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 41818
Number of edges: 1067042

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9496
Number of communities: 24
Elapsed time: 2 seconds</code></pre>
</div>
</div>
<p>Let’s visualize the BANKSY clusters alongside the Seurat clusters for a side-by-side comparison:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>banksy_cluster)),</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>banksy_cluster))</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>image_banksy_clusters <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(object_filt, <span class="at">group.by =</span> <span class="st">"banksy_cluster"</span>,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">pt.size.factor =</span> <span class="dv">7</span>, <span class="at">cols =</span> color_pal)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters <span class="sc">|</span> image_banksy_clusters</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can see that, as expected, the BANKSY clusters are more spatially-restricted, or more compact, than the Seurat clusters. We also see that the BANKSY clusters are less noisy than the Seurat clusters, likely because of the smoothing effect of considering a cell’s spatial neighborhood when assigning a cluster label.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click here to see BANKSY using a lambda value of 0.2
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If we had run BANKSY with <code>lambda = 0.2</code>, as recommended for cell type clustering instead of <code>lambda = 0.8</code> for spatial domain clustering, the resultant clusters would be less spatially restricted (in other words less compact and more distributed throughout the image) and more similar to our Seurat clustering. Below is a figure using <code>lamba=0.2</code> in BANKSY rather than <code>lamba=0.8</code>.</p>
<p align="center">
<img src="../img/banksy_clustering_lambda_0.2.png" width="450">
</p>
</div>
</div>
</div>
</section>
<section id="cell-type-annotation" class="level2">
<h2 class="anchored" data-anchor-id="cell-type-annotation">Cell Type Annotation</h2>
<p>Perhaps we are particularly interested in understanding the organization of cell types in the cortical region of the brain.</p>
<p align="center">
<img src="../img/Cell_type_annotations.png" width="800">
</p>
<p>We first subset our Seurat object to this region of interest. There are multiple ways of subsetting a Seurat object to a region of interest, but here we have identified a handful of cluster numbers that appear almost exclusively in the cortical region, and we will subset the object to only include cells that are assigned these cluster numbers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">subset</span>(object_filt, seurat_cluster.projected <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">18</span>, <span class="dv">19</span>, <span class="dv">7</span>, <span class="dv">2</span>, <span class="dv">4</span>))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected)),</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected))</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(cortex, <span class="at">group.by =</span> <span class="st">'seurat_cluster.projected'</span>, </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">pt.size.factor =</span> <span class="dv">8</span>, <span class="at">cols =</span> color_pal)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note: Your colors may be different than the ones in the above figure.</p>
</div>
</div>
<p>To perform accurate annotation of cell types, we must also take into consideration that our 16µm x 16µm bins may contain one or more cells each. The method <a href="https://www.nature.com/articles/s41587-021-00830-w">Robust Cell Type Deconvolution</a> (RCTD) has been shown to accurately annotate spatial data from a variety of technologies while taking into consideration that a single bin may exhibit multiple cell type profiles.</p>
<p>RCTD takes a cell-type-annotated scRNA-seq dataset as a reference and a spatial dataset as a query. For our reference, we will use a subsampled version of the mouse scRNA-seq dataset from the Allen Brain Atlas. We will use our cortex Seurat object as the spatial query. As an overview, the process is as follows:</p>
<ol type="1">
<li>Sketch and process the spatial query dataset</li>
<li>Load and format the scRNA-seq reference dataset</li>
<li>Apply RCTD to deconvolute the “sketched” cortical cells and annotate them</li>
<li>Project these annotations onto the full cortical dataset.</li>
</ol>
<section id="sketch-and-process-the-spatial-query-dataset" class="level3">
<h3 class="anchored" data-anchor-id="sketch-and-process-the-spatial-query-dataset">1) Sketch and process the spatial query dataset</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">'Spatial.016um'</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cortex)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">SketchData</span>(</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncells =</span> <span class="dv">3000</span>,</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"LeverageScore"</span>,</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"sketch"</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cortex)</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cortex, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">reduction.name =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">verbose =</span> T)</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, </span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>                  <span class="at">reduction.name =</span> <span class="st">"umap.cortex.sketch"</span>, <span class="at">return.model =</span> T, </span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, <span class="at">verbose =</span> T)</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="co"># create the RCTD query object</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>counts_hd <span class="ot">&lt;-</span> cortex[[<span class="st">"sketch"</span>]]<span class="sc">$</span>counts</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>cortex_cells_hd <span class="ot">&lt;-</span> <span class="fu">colnames</span>(cortex[[<span class="st">"sketch"</span>]])</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(cortex)[cortex_cells_hd, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">SpatialRNA</span>(coords, counts_hd, <span class="fu">colSums</span>(counts_hd))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-and-format-the-reference-dataset" class="level3">
<h3 class="anchored" data-anchor-id="load-and-format-the-reference-dataset">2) Load and format the reference dataset</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mem.maxVSize</span>(<span class="dv">15000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ref_subset <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"data_processed/allen_scRNAseq_ref_subset.qs"</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ref_subset) <span class="ot">&lt;-</span> <span class="st">"subclass_label"</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> ref_subset[[<span class="st">"RNA"</span>]]<span class="sc">$</span>counts</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(ref_subset<span class="sc">$</span>subclass_label)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>nUMI <span class="ot">&lt;-</span> ref_subset<span class="sc">$</span>nCount_RNA</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(cluster) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"/"</span>, <span class="st">"-"</span>, <span class="fu">levels</span>(cluster))</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">droplevels</span>(cluster)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create the RCTD reference object</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">Reference</span>(counts, cluster, nUMI)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="apply-rctd-to-deconvolute-the-sketched-cortical-cells-and-annotate-them" class="level3">
<h3 class="anchored" data-anchor-id="apply-rctd-to-deconvolute-the-sketched-cortical-cells-and-annotate-them">3) Apply RCTD to deconvolute the “sketched” cortical cells and annotate them</h3>
<p>Note that <code>run.RCTD</code> takes 10-15 minutes to complete on a laptop using 6 cores</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run RCTD</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">create.RCTD</span>(query, reference, <span class="at">max_cores =</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
          Astro        CA1-ProS       CA2-IG-FC             CA3            Car3 
            500             500             101             442             500 
             CR          CT SUB              DG            Endo      L2 IT ENTl 
            208             500             500             500             500 
     L2 IT ENTm     L2-3 IT CTX    L2-3 IT ENTl     L2-3 IT PPP     L2-3 IT RHP 
            433             500             500             500             500 
      L3 IT ENT      L4 RSP-ACA     L4-5 IT CTX       L5 IT CTX          L5 PPP 
            500             500             500             500             254 
      L5 PT CTX L5-6 IT TPE-ENT     L5-6 NP CTX       L6 CT CTX       L6 IT CTX 
            500             500             500             500             500 
     L6 IT ENTl         L6b CTX      L6b-CT ENT           Lamp5       Micro-PVM 
            273             500             500             500             500 
         NP PPP          NP SUB           Oligo           Pvalb        SMC-Peri 
            500             398             500             500             285 
           Sncg             Sst       Sst Chodl        SUB-ProS             Vip 
            500             500             495             500             500 
           VLMC 
            152 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">run.RCTD</span>(RCTD, <span class="at">doublet_mode =</span> <span class="st">"doublet"</span>) <span class="co"># this command takes ~15 mins to run</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gather_results: finished 1000"
[1] "gather_results: finished 2000"
[1] "gather_results: finished 3000"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add results back to Seurat object</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(cortex, <span class="at">metadata =</span> RCTD<span class="sc">@</span>results<span class="sc">$</span>results_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="project-rctd-labels-onto-all-cortical-cells" class="level3">
<h3 class="anchored" data-anchor-id="project-rctd-labels-onto-all-cortical-cells">4) Project RCTD labels onto all cortical cells</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type <span class="ot">&lt;-</span> <span class="fu">as.character</span>(cortex<span class="sc">$</span>first_type)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type[<span class="fu">is.na</span>(cortex<span class="sc">$</span>first_type)] <span class="ot">&lt;-</span> <span class="st">"Unknown"</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ProjectData</span>(</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"Spatial.016um"</span>,</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.reduction =</span> <span class="st">"pca.cortex"</span>,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span>,</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.reduction =</span> <span class="st">"pca.cortex.sketch"</span>,</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.model =</span> <span class="st">"umap.cortex.sketch"</span>,</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">full_first_type =</span> <span class="st">"first_type"</span>)</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that the excitatory neurons are located in layers at varying cortical depths, as expected.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"full_first_type"</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">CellsByIdentities</span>(cortex)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Layered (starts with L), excitatory neurons in the cortex</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>excitatory_names <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">grep</span>(<span class="st">"^L.* CTX"</span>, <span class="fu">names</span>(cells), <span class="at">value =</span> <span class="cn">TRUE</span>))</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(cortex, <span class="at">cells.highlight =</span> cells[excitatory_names], </span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">cols.highlight =</span> <span class="fu">c</span>(<span class="st">"#FFFF00"</span>, <span class="st">"grey50"</span>), <span class="at">facet.highlight =</span> T, </span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">combine =</span> T, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">pt.size.factor =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="visium_hd_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="save-r-session" class="level1">
<h1>Save R Session</h1>
<p>At the end of any analysis, it is important to save your session information! This will let future users know which versions of packages are necessary in order to reproduce the same results.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
sessionInfo
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.3.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Detroit
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] future_1.67.0        DT_0.34.0            spacexr_2.2.1       
 [4] quadprog_1.5-8       Banksy_1.2.0         SeuratWrappers_0.4.0
 [7] qs_0.27.3            Seurat_5.3.1         SeuratObject_5.2.0  
[10] sp_2.2-0             patchwork_1.3.2      lubridate_1.9.4     
[13] forcats_1.0.1        stringr_1.6.0        dplyr_1.1.4         
[16] purrr_1.2.0          readr_2.1.5          tidyr_1.3.1         
[19] tibble_3.3.0         ggplot2_4.0.0        tidyverse_2.0.0     

loaded via a namespace (and not attached):
  [1] RcppHungarian_0.3           RcppAnnoy_0.0.22           
  [3] splines_4.4.1               later_1.4.4                
  [5] R.oo_1.27.1                 polyclip_1.10-7            
  [7] fastDummies_1.7.5           lifecycle_1.0.4            
  [9] aricode_1.0.3               doParallel_1.0.17          
 [11] globals_0.18.0              lattice_0.22-7             
 [13] MASS_7.3-65                 crosstalk_1.2.2            
 [15] magrittr_2.0.4              sass_0.4.10                
 [17] plotly_4.11.0               rmarkdown_2.30             
 [19] jquerylib_0.1.4             yaml_2.3.10                
 [21] remotes_2.5.0               httpuv_1.6.16              
 [23] otel_0.2.0                  sctransform_0.4.2          
 [25] spam_2.11-1                 spatstat.sparse_3.1-0      
 [27] reticulate_1.44.0           cowplot_1.2.0              
 [29] pbapply_1.7-4               RColorBrewer_1.1-3         
 [31] abind_1.4-8                 zlibbioc_1.52.0            
 [33] Rtsne_0.17                  GenomicRanges_1.58.0       
 [35] R.utils_2.13.0              BiocGenerics_0.52.0        
 [37] GenomeInfoDbData_1.2.13     IRanges_2.40.1             
 [39] S4Vectors_0.44.0            ggrepel_0.9.6              
 [41] irlba_2.3.5.1               listenv_0.10.0             
 [43] spatstat.utils_3.2-0        goftest_1.2-3              
 [45] RSpectra_0.16-2             spatstat.random_3.4-2      
 [47] fitdistrplus_1.2-4          parallelly_1.45.1          
 [49] codetools_0.2-20            DelayedArray_0.32.0        
 [51] RApiSerialize_0.1.4         tidyselect_1.2.1           
 [53] UCSC.utils_1.2.0            farver_2.1.2               
 [55] matrixStats_1.5.0           stats4_4.4.1               
 [57] spatstat.explore_3.5-3      jsonlite_2.0.0             
 [59] progressr_0.18.0            iterators_1.0.14           
 [61] ggridges_0.5.7              survival_3.8-3             
 [63] foreach_1.5.2               dbscan_1.2.3               
 [65] tools_4.4.1                 ica_1.0-3                  
 [67] Rcpp_1.1.0                  glue_1.8.0                 
 [69] gridExtra_2.3               SparseArray_1.6.2          
 [71] xfun_0.54                   MatrixGenerics_1.18.1      
 [73] GenomeInfoDb_1.42.3         withr_3.0.2                
 [75] BiocManager_1.30.26         fastmap_1.2.0              
 [77] digest_0.6.38               rsvd_1.0.5                 
 [79] timechange_0.3.0            R6_2.6.1                   
 [81] mime_0.13                   scattermore_1.2            
 [83] sccore_1.0.6                tensor_1.5.1               
 [85] dichromat_2.0-0.1           spatstat.data_3.1-9        
 [87] R.methodsS3_1.8.2           generics_0.1.4             
 [89] data.table_1.17.8           httr_1.4.7                 
 [91] htmlwidgets_1.6.4           S4Arrays_1.6.0             
 [93] uwot_0.2.4                  pkgconfig_2.0.3            
 [95] gtable_0.3.6                lmtest_0.9-40              
 [97] S7_0.2.0                    SingleCellExperiment_1.28.1
 [99] XVector_0.46.0              htmltools_0.5.8.1          
[101] dotCall64_1.2               scales_1.4.0               
[103] Biobase_2.66.0              png_0.1-8                  
[105] SpatialExperiment_1.16.0    spatstat.univar_3.1-4      
[107] knitr_1.50                  tzdb_0.5.0                 
[109] reshape2_1.4.5              rjson_0.2.23               
[111] nlme_3.1-168                cachem_1.1.0               
[113] zoo_1.8-14                  KernSmooth_2.23-26         
[115] vipor_0.4.7                 parallel_4.4.1             
[117] miniUI_0.1.2                ggrastr_1.0.2              
[119] pillar_1.11.1               grid_4.4.1                 
[121] vctrs_0.6.5                 RANN_2.6.2                 
[123] promises_1.5.0              stringfish_0.17.0          
[125] xtable_1.8-4                cluster_2.1.8.1            
[127] beeswarm_0.4.0              evaluate_1.0.5             
[129] magick_2.9.0                cli_3.6.5                  
[131] compiler_4.4.1              rlang_1.1.6                
[133] crayon_1.5.3                future.apply_1.20.0        
[135] labeling_0.4.3              mclust_6.1.2               
[137] ggbeeswarm_0.7.2            plyr_1.8.9                 
[139] stringi_1.8.7               viridisLite_0.4.2          
[141] deldir_2.0-4                lazyeval_0.2.2             
[143] spatstat.geom_3.6-0         Matrix_1.7-4               
[145] RcppHNSW_0.6.0              hms_1.1.4                  
[147] shiny_1.11.1                SummarizedExperiment_1.36.0
[149] ROCR_1.0-11                 leidenAlg_1.1.5            
[151] igraph_2.2.1                bslib_0.9.0                
[153] RcppParallel_5.1.11-1      </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb55" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Visium HD Analysis"</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="an">authors:</span><span class="co"> "Alex Bartlett, Meeta Mistry, and Will Gammerdinger"</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "February 26th, 2025"</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co">  warning: false</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>Contributors: Alex Bartlett, Meeta Mistry and Will Gammerdinger</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>Approximate time: 2 hours and 45 minutes</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a><span class="fu"># Learning Objectives </span></span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>In this lesson, we will:</span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Describe the elements of the Seurat object and how they are generated</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Visually inspect and compare spatial scRNA-seq data before and after filtering</span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Execute clustering workflows and visualize results on a tissue section</span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a><span class="fu"># NGS-based Spatial Transcriptomics Data Analysis</span></span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-24"><a href="#cb55-24" aria-hidden="true" tabindex="-1"></a><span class="fu">## Mouse Brain Visium HD </span></span>
<span id="cb55-25"><a href="#cb55-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-26"><a href="#cb55-26" aria-hidden="true" tabindex="-1"></a>The Visium HD platform is compatible with human and mouse fresh frozen, fixed frozen, and formalin-fixed parrafin-embedded (FFPE) tissue sections. For this lesson, we will be working with data from a fresh frozen coronal section of a mouse brain sample.</span>
<span id="cb55-27"><a href="#cb55-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-28"><a href="#cb55-28" aria-hidden="true" tabindex="-1"></a>Each Visium HD slide has the same 6.5 x 6.5mm capture area as previous Visium products but is covered with about 11 million tiles. These 2µm x 2µm squares are arrayed in a continuous lawn across the entire capture area. The squares are each uniquely barcoded with an oligonucleotide and contain probes allowing for the detection of the full coding transcriptome. </span>
<span id="cb55-29"><a href="#cb55-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-30"><a href="#cb55-30" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-31"><a href="#cb55-31" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/visium_hd_slide_graphic.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800"</span><span class="dt">&gt;</span></span>
<span id="cb55-32"><a href="#cb55-32" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-33"><a href="#cb55-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-34"><a href="#cb55-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-35"><a href="#cb55-35" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preprocessing Data with Spaceranger</span></span>
<span id="cb55-36"><a href="#cb55-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-37"><a href="#cb55-37" aria-hidden="true" tabindex="-1"></a>Sequencing facilities often output scRNA-seq data, including spatial scRNA-seq data, in FASTQ format. Because this is Visium HD data from 10X Genomics, we used their proprietary preprocessing software <span class="co">[</span><span class="ot">Space Ranger</span><span class="co">](https://www.10xgenomics.com/support/software/space-ranger/latest)</span> to process the FASTQ files into a count matrix and other images. Specifically, the ```spaceranger count``` command aligns the reads in the FASTQ files against a transcriptomic reference and provides their spatial location using the oligonucleotide barcode. </span>
<span id="cb55-38"><a href="#cb55-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-39"><a href="#cb55-39" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse=true}</span>
<span id="cb55-40"><a href="#cb55-40" aria-hidden="true" tabindex="-1"></a><span class="fu"># Click here to see an example of how to run `spaceranger count`</span></span>
<span id="cb55-41"><a href="#cb55-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-42"><a href="#cb55-42" aria-hidden="true" tabindex="-1"></a>A sample command for running <span class="in">`spaceranger count`</span> is:</span>
<span id="cb55-43"><a href="#cb55-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-46"><a href="#cb55-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb55-47"><a href="#cb55-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb55-48"><a href="#cb55-48" aria-hidden="true" tabindex="-1"></a><span class="ex">spaceranger</span> count <span class="at">--id</span><span class="op">=</span>hd_count <span class="dt">\</span></span>
<span id="cb55-49"><a href="#cb55-49" aria-hidden="true" tabindex="-1"></a>   <span class="at">--transcriptome</span><span class="op">=</span>/path/to/refdata-gex-GRCh38-2020-A <span class="dt">\</span></span>
<span id="cb55-50"><a href="#cb55-50" aria-hidden="true" tabindex="-1"></a>   <span class="at">--fastqs</span><span class="op">=</span>/path/to/fastq <span class="dt">\</span></span>
<span id="cb55-51"><a href="#cb55-51" aria-hidden="true" tabindex="-1"></a>   <span class="at">--probe-set</span><span class="op">=</span>/path/to/Visium_Human_Transcriptome_Probe_Set_v2.0_GRCh38-2020-A.csv <span class="dt">\</span></span>
<span id="cb55-52"><a href="#cb55-52" aria-hidden="true" tabindex="-1"></a>   <span class="at">--slide</span><span class="op">=</span>H1-YD7CDZK <span class="dt">\</span></span>
<span id="cb55-53"><a href="#cb55-53" aria-hidden="true" tabindex="-1"></a>   <span class="at">--area</span><span class="op">=</span>A1 <span class="dt">\</span></span>
<span id="cb55-54"><a href="#cb55-54" aria-hidden="true" tabindex="-1"></a>   <span class="at">--cytaimage</span><span class="op">=</span>/path/to/CAVG10539_2023-11-16_14-56-24_APPS115_H1-YD7CDZK_A1_S11088.tif <span class="dt">\</span></span>
<span id="cb55-55"><a href="#cb55-55" aria-hidden="true" tabindex="-1"></a>   <span class="at">--image</span><span class="op">=</span>/path/to/APPS115_11088_rescan_01.btf <span class="dt">\</span></span>
<span id="cb55-56"><a href="#cb55-56" aria-hidden="true" tabindex="-1"></a>   <span class="at">--create-bam</span><span class="op">=</span>false</span>
<span id="cb55-57"><a href="#cb55-57" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-58"><a href="#cb55-58" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-59"><a href="#cb55-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-60"><a href="#cb55-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-61"><a href="#cb55-61" aria-hidden="true" tabindex="-1"></a>Note that Space Ranger requires a Linux system with at least 32 cores, 64GB of RAM, and 1TB of disk space. </span>
<span id="cb55-62"><a href="#cb55-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-63"><a href="#cb55-63" aria-hidden="true" tabindex="-1"></a>When ```spaceranger count``` completes successfully, it will generate output similar to the following, which will enable the analyst to perform further analysis in R or Python or using the proprietary Loupe browser from 10X Genomics. </span>
<span id="cb55-64"><a href="#cb55-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-65"><a href="#cb55-65" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-66"><a href="#cb55-66" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/spaceranger_output.svg"</span><span class="ot"> alt</span><span class="op">=</span><span class="st">"spaceranger output"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"600"</span><span class="dt">/&gt;</span></span>
<span id="cb55-67"><a href="#cb55-67" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-68"><a href="#cb55-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-69"><a href="#cb55-69" aria-hidden="true" tabindex="-1"></a>In the Visium HD assay, in addition to providing data at the level of the 2µm x 2µm squares, Space Ranger also bins the 2µm x 2µm squares into 8µm x 8µm and 16µm x 16µm bins. Most of the above output is produced for each binning resolution. </span>
<span id="cb55-70"><a href="#cb55-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-71"><a href="#cb55-71" aria-hidden="true" tabindex="-1"></a>While the single-digit micron resolution is a big technological improvement over original Visium’s original ∼55μm spots, the higher resolution also presents challenges. Having access to 2μm bins along with matching morphology information provides a great opportunity for reconstructing single cells from the data, which is undoubtedly very powerful. However, because the 2µm  x 2µm squares (and even the 8µm x 8µm bins) are so small, there is a potential for very little biological signal to be captured per bin. Additionally, the sheer number of bins at these higher resolutions can present challenges in terms of computational time and resources. **For the purposes of this lesson, we will use the 16µm x 16µm bins.**</span>
<span id="cb55-72"><a href="#cb55-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-73"><a href="#cb55-73" aria-hidden="true" tabindex="-1"></a>We can view and explore the web summary HTML of our data found in the "reports" folder of your project. </span>
<span id="cb55-74"><a href="#cb55-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-75"><a href="#cb55-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-76"><a href="#cb55-76" aria-hidden="true" tabindex="-1"></a><span class="fu">## Analysis workflow</span></span>
<span id="cb55-77"><a href="#cb55-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-78"><a href="#cb55-78" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-79"><a href="#cb55-79" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Full_workflow.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800"</span><span class="dt">&gt;</span></span>
<span id="cb55-80"><a href="#cb55-80" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-81"><a href="#cb55-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-82"><a href="#cb55-82" aria-hidden="true" tabindex="-1"></a><span class="fu">### Setting up </span></span>
<span id="cb55-83"><a href="#cb55-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-84"><a href="#cb55-84" aria-hidden="true" tabindex="-1"></a>For this module, we will be working within an RStudio project. In order to follow along you should have **downloaded the R project**.</span>
<span id="cb55-85"><a href="#cb55-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-86"><a href="#cb55-86" aria-hidden="true" tabindex="-1"></a>::: callout-important</span>
<span id="cb55-87"><a href="#cb55-87" aria-hidden="true" tabindex="-1"></a>If you haven't done this already, the project is located in "Dataset for workshop" -&gt; "Day 2- NGS-based- VisiumHD" in the course DropBox.</span>
<span id="cb55-88"><a href="#cb55-88" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-89"><a href="#cb55-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-90"><a href="#cb55-90" aria-hidden="true" tabindex="-1"></a>Once downloaded, you should see a file called <span class="in">`visiumHD_nanocourse.zip`</span> on your computer (likely, in your <span class="in">`Downloads`</span> folder). </span>
<span id="cb55-91"><a href="#cb55-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-92"><a href="#cb55-92" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Unzip this file.** This will result in a folder of the same name. </span>
<span id="cb55-93"><a href="#cb55-93" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Move the folder to the location on your computer where you would like to perform the analysis.**</span>
<span id="cb55-94"><a href="#cb55-94" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Open up the folder.** The contents will look like the screenshot below:</span>
<span id="cb55-95"><a href="#cb55-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-96"><a href="#cb55-96" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-97"><a href="#cb55-97" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/proj_screenshot.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"500"</span><span class="dt">&gt;</span></span>
<span id="cb55-98"><a href="#cb55-98" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-99"><a href="#cb55-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-100"><a href="#cb55-100" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Locate the `.Rproj file` and double-click on it.** This will open up RStudio with the "visiumHD_nanocourse" project loaded. </span>
<span id="cb55-101"><a href="#cb55-101" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>**Open a new Rscript file.**</span>
<span id="cb55-102"><a href="#cb55-102" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>**Start with some comments to indicate what this file is going to contain:**</span>
<span id="cb55-103"><a href="#cb55-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-106"><a href="#cb55-106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-107"><a href="#cb55-107" aria-hidden="true" tabindex="-1"></a><span class="co"># February 26th, 2025</span></span>
<span id="cb55-108"><a href="#cb55-108" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatial Transcriptomics: Session 2</span></span>
<span id="cb55-109"><a href="#cb55-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-110"><a href="#cb55-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-111"><a href="#cb55-111" aria-hidden="true" tabindex="-1"></a><span class="ss">7. </span>**Save the Rscript in the `code` folder as `visiumHD.R`.** Your working directory should look something like this:</span>
<span id="cb55-112"><a href="#cb55-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-113"><a href="#cb55-113" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-114"><a href="#cb55-114" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Code_folder.gif"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-115"><a href="#cb55-115" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-116"><a href="#cb55-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-117"><a href="#cb55-117" aria-hidden="true" tabindex="-1"></a><span class="fu">## Loading Libraries</span></span>
<span id="cb55-118"><a href="#cb55-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-119"><a href="#cb55-119" aria-hidden="true" tabindex="-1"></a>Next, we will need to be sure to load the libraries that we will be using:</span>
<span id="cb55-120"><a href="#cb55-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-123"><a href="#cb55-123" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-124"><a href="#cb55-124" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb55-125"><a href="#cb55-125" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb55-126"><a href="#cb55-126" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb55-127"><a href="#cb55-127" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb55-128"><a href="#cb55-128" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span>
<span id="cb55-129"><a href="#cb55-129" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb55-130"><a href="#cb55-130" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Banksy)</span>
<span id="cb55-131"><a href="#cb55-131" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(quadprog)</span>
<span id="cb55-132"><a href="#cb55-132" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spacexr)</span>
<span id="cb55-133"><a href="#cb55-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-134"><a href="#cb55-134" aria-hidden="true" tabindex="-1"></a><span class="co"># Increases the size of the default vector</span></span>
<span id="cb55-135"><a href="#cb55-135" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize=</span> <span class="dv">2000000000</span>)</span>
<span id="cb55-136"><a href="#cb55-136" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-137"><a href="#cb55-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-138"><a href="#cb55-138" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating the Seurat Object</span></span>
<span id="cb55-139"><a href="#cb55-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-140"><a href="#cb55-140" aria-hidden="true" tabindex="-1"></a>The Seurat object is a custom list-like object that has well-defined spaces to store specific information/data for single-cell experiments, including spatial experiments and Visium HD.</span>
<span id="cb55-141"><a href="#cb55-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-142"><a href="#cb55-142" aria-hidden="true" tabindex="-1"></a>The Seurat package provides a function ```Load10X_Spatial()``` to easily create a Seurat object from the output of Space Ranger. The ```Load10X_Spatial``` function takes as input the feature matrix and the low-resolution tissue image from the output of Space Ranger and generates a Seurat object containing both gene-level counts and spatial information. </span>
<span id="cb55-143"><a href="#cb55-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-144"><a href="#cb55-144" aria-hidden="true" tabindex="-1"></a>**We will not have you run this code**, as this can take some time and the Space Ranger output files are quite large to share. Instead, you will load the pre-made Seurat object. </span>
<span id="cb55-145"><a href="#cb55-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-146"><a href="#cb55-146" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse=true}</span>
<span id="cb55-147"><a href="#cb55-147" aria-hidden="true" tabindex="-1"></a><span class="fu"># Click here to see the R code used to create the Seurat object</span></span>
<span id="cb55-148"><a href="#cb55-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-151"><a href="#cb55-151" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-152"><a href="#cb55-152" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb55-153"><a href="#cb55-153" aria-hidden="true" tabindex="-1"></a>localdir <span class="ot">&lt;-</span> <span class="st">'../path/to/spaceranger/outs/'</span></span>
<span id="cb55-154"><a href="#cb55-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-155"><a href="#cb55-155" aria-hidden="true" tabindex="-1"></a><span class="co">#Load the raw feature matrix</span></span>
<span id="cb55-156"><a href="#cb55-156" aria-hidden="true" tabindex="-1"></a>object <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(<span class="at">data.dir =</span> localdir,</span>
<span id="cb55-157"><a href="#cb55-157" aria-hidden="true" tabindex="-1"></a>                          <span class="at">filename =</span> <span class="st">'raw_feature_bc_matrix.h5'</span>,</span>
<span id="cb55-158"><a href="#cb55-158" aria-hidden="true" tabindex="-1"></a>                          <span class="at">bin.size =</span> <span class="dv">16</span>)</span>
<span id="cb55-159"><a href="#cb55-159" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-160"><a href="#cb55-160" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-161"><a href="#cb55-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-162"><a href="#cb55-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-163"><a href="#cb55-163" aria-hidden="true" tabindex="-1"></a><span class="fu">### Explore the object</span></span>
<span id="cb55-164"><a href="#cb55-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-165"><a href="#cb55-165" aria-hidden="true" tabindex="-1"></a>Let's read in the Seurat object and talk about some very basic slots that we will be accessing. </span>
<span id="cb55-166"><a href="#cb55-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-169"><a href="#cb55-169" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-170"><a href="#cb55-170" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in Seurat object</span></span>
<span id="cb55-171"><a href="#cb55-171" aria-hidden="true" tabindex="-1"></a>object <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">'data_processed/MsBrain_FF-A1_subset_misc.qs'</span>)</span>
<span id="cb55-172"><a href="#cb55-172" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-173"><a href="#cb55-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-174"><a href="#cb55-174" aria-hidden="true" tabindex="-1"></a>We can print the Seurat object by using:</span>
<span id="cb55-175"><a href="#cb55-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-178"><a href="#cb55-178" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-179"><a href="#cb55-179" aria-hidden="true" tabindex="-1"></a>object</span>
<span id="cb55-180"><a href="#cb55-180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-181"><a href="#cb55-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-182"><a href="#cb55-182" aria-hidden="true" tabindex="-1"></a>Now we can examine its major features, which we will add to and alter throughout the lesson:</span>
<span id="cb55-183"><a href="#cb55-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-184"><a href="#cb55-184" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-185"><a href="#cb55-185" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Base_seurat_object_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"1000"</span><span class="dt">&gt;</span></span>
<span id="cb55-186"><a href="#cb55-186" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-187"><a href="#cb55-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-188"><a href="#cb55-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-189"><a href="#cb55-189" aria-hidden="true" tabindex="-1"></a>:::{.callout-tip}</span>
<span id="cb55-190"><a href="#cb55-190" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercise</span></span>
<span id="cb55-191"><a href="#cb55-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-192"><a href="#cb55-192" aria-hidden="true" tabindex="-1"></a>There are 3 things about our Seurat object printout that would be different if we were using the 8µm x 8µm binning instead of the 16µm x 16µm binning. What are these three differences?</span>
<span id="cb55-193"><a href="#cb55-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-194"><a href="#cb55-194" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse=true}</span>
<span id="cb55-195"><a href="#cb55-195" aria-hidden="true" tabindex="-1"></a><span class="fu"># Click here to use an app that lets you explore different bin sizes for this Seurat object</span></span>
<span id="cb55-196"><a href="#cb55-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-197"><a href="#cb55-197" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;&lt;</span><span class="kw">iframe</span><span class="ot"> src</span><span class="op">=</span><span class="st">"https://hcbc.connect.hms.harvard.edu/Spatial_resolution_question_1/?showcase=0"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"600px"</span><span class="ot"> height</span><span class="op">=</span><span class="st">"250px"</span><span class="ot"> data-external</span><span class="op">=</span><span class="st">"1"</span><span class="dt">&gt;</span> <span class="dt">&lt;/</span><span class="kw">iframe</span><span class="dt">&gt;&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-198"><a href="#cb55-198" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-199"><a href="#cb55-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-200"><a href="#cb55-200" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-201"><a href="#cb55-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-202"><a href="#cb55-202" aria-hidden="true" tabindex="-1"></a><span class="fu">## Quality Control</span></span>
<span id="cb55-203"><a href="#cb55-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-204"><a href="#cb55-204" aria-hidden="true" tabindex="-1"></a>The main objective of quality control is to filter the data so that we include only data from bins that are of high-quality. This makes it so that when we cluster our bins, it is easier to identify distinct cell type populations.</span>
<span id="cb55-205"><a href="#cb55-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-206"><a href="#cb55-206" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-207"><a href="#cb55-207" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/QC_workflow.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800"</span><span class="dt">&gt;</span></span>
<span id="cb55-208"><a href="#cb55-208" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-209"><a href="#cb55-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-210"><a href="#cb55-210" aria-hidden="true" tabindex="-1"></a>In Visium HD data, the main challenge is in **delineating bins that are poor quality from bins containing reads from less complex cells**. If you expect a particular cell type in your dataset to be less transcriptionally active as compared other cell types in your dataset, the bins underneath this cell type will naturally have fewer detected genes and transcripts. However, having fewer detected genes and transcripts can also be a technical artifact and not a result of biological signal. </span>
<span id="cb55-211"><a href="#cb55-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-212"><a href="#cb55-212" aria-hidden="true" tabindex="-1"></a>Various metrics can be used to filter low-quality bins from high-quality ones, including:</span>
<span id="cb55-213"><a href="#cb55-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-214"><a href="#cb55-214" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**UMI counts per bin** - This is the number of unique transcripts detected per bin. Because the bins are very small, this number is less than what we would expect for non-spatial scRNA-seq data.</span>
<span id="cb55-215"><a href="#cb55-215" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Genes detected per bin** - This is the number of unique genes detected per bin. Again, because the bins are very small, this number is less than what we would expect for non-spatial scRNA-seq data.</span>
<span id="cb55-216"><a href="#cb55-216" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Complexity (novelty score)** - The novelty score is computed as shown below:</span>
<span id="cb55-217"><a href="#cb55-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-218"><a href="#cb55-218" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-219"><a href="#cb55-219" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"https://latex.codecogs.com/svg.image?\text{Complexity</span><span class="dv">&amp;space;</span><span class="st">Score}=\frac{\text{Number</span><span class="dv">&amp;space;</span><span class="st">of</span><span class="dv">&amp;space;</span><span class="st">Genes}}{\text{Number</span><span class="dv">&amp;space;</span><span class="st">of</span><span class="dv">&amp;space;</span><span class="st">UMIs}}"</span><span class="ot"> </span><span class="dt">/&gt;</span></span>
<span id="cb55-220"><a href="#cb55-220" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-221"><a href="#cb55-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-222"><a href="#cb55-222" aria-hidden="true" tabindex="-1"></a>   If there are many captured transcripts (high nUMI) and a low number of genes detected in a bin, this likely means that you only captured a low number of genes and simply sequenced transcripts from those lower number of genes over and over again. These low complexity (low novelty) bins could represent a specific cell type (i.e. red blood cells, which lack a typical transcriptome), or could be due to an artifact or contamination. Generally, we expect the complexity score to be above 0.80 for good-quality bins.</span>
<span id="cb55-223"><a href="#cb55-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-224"><a href="#cb55-224" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Mitochondrial counts ratio** - This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells. We define poor-quality samples for mitochondrial counts as bins which surpass the 0.2 mitochondrial ratio threshold, unless of course you are expecting this in your sample. This ratio is computed as:</span>
<span id="cb55-225"><a href="#cb55-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-226"><a href="#cb55-226" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-227"><a href="#cb55-227" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"https://latex.codecogs.com/svg.image?\text{Mitochondiral</span><span class="dv">&amp;space;</span><span class="st">Ratio}=\frac{\text{Number</span><span class="dv">&amp;space;</span><span class="st">of</span><span class="dv">&amp;space;</span><span class="st">reads</span><span class="dv">&amp;space;</span><span class="st">aligning</span><span class="dv">&amp;space;</span><span class="st">to</span><span class="dv">&amp;space;</span><span class="st">mitochondrial</span><span class="dv">&amp;space;</span><span class="st">genes}}{\text{Total</span><span class="dv">&amp;space;</span><span class="st">reads}}"</span><span class="ot"> </span><span class="dt">/&gt;</span></span>
<span id="cb55-228"><a href="#cb55-228" aria-hidden="true" tabindex="-1"></a>   <span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-229"><a href="#cb55-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-230"><a href="#cb55-230" aria-hidden="true" tabindex="-1"></a>Let's take a quick look at the data and make a decision on whether we need to apply any filtering. We will examine the distributions of UMI counts per bin and genes detected per bin to determine reasonable thresholds for those metrics to implement during QC filtering.</span>
<span id="cb55-231"><a href="#cb55-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-232"><a href="#cb55-232" aria-hidden="true" tabindex="-1"></a><span class="fu">### Pre-filtering </span></span>
<span id="cb55-233"><a href="#cb55-233" aria-hidden="true" tabindex="-1"></a>To create some plots first, we will need to create a metadata object using this command:</span>
<span id="cb55-234"><a href="#cb55-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-237"><a href="#cb55-237" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-238"><a href="#cb55-238" aria-hidden="true" tabindex="-1"></a>object_meta <span class="ot">&lt;-</span> object<span class="sc">@</span>meta.data</span>
<span id="cb55-239"><a href="#cb55-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-240"><a href="#cb55-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-241"><a href="#cb55-241" aria-hidden="true" tabindex="-1"></a>Now we can plot the number of UMIs (nUMI) and the number of genes (nGene) side-by-side. For both of the plots, we expect to see a bimodal distribution, with one peak representing bins containing lower-quality cells with fewer genes and UMIs and another peak representing bins containing healthy cells with more genes and UMIs. Ideally, the peak representing lower-quality and dying cells is small and the peak representing healthy cells is large.</span>
<span id="cb55-242"><a href="#cb55-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-245"><a href="#cb55-245" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-246"><a href="#cb55-246" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot for nUMI</span></span>
<span id="cb55-247"><a href="#cb55-247" aria-hidden="true" tabindex="-1"></a>dist_counts_before <span class="ot">&lt;-</span> object_meta <span class="sc">%&gt;%</span></span>
<span id="cb55-248"><a href="#cb55-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nCount_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb55-249"><a href="#cb55-249" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb55-250"><a href="#cb55-250" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb55-251"><a href="#cb55-251" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb55-252"><a href="#cb55-252" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb55-253"><a href="#cb55-253" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of UMIs per bin"</span>) <span class="sc">+</span></span>
<span id="cb55-254"><a href="#cb55-254" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Pre-QC UMIs/Bin'</span>) <span class="sc">+</span></span>
<span id="cb55-255"><a href="#cb55-255" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb55-256"><a href="#cb55-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-257"><a href="#cb55-257" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot for nGene</span></span>
<span id="cb55-258"><a href="#cb55-258" aria-hidden="true" tabindex="-1"></a>dist_features_before <span class="ot">&lt;-</span> object_meta <span class="sc">%&gt;%</span></span>
<span id="cb55-259"><a href="#cb55-259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nFeature_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb55-260"><a href="#cb55-260" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb55-261"><a href="#cb55-261" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb55-262"><a href="#cb55-262" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb55-263"><a href="#cb55-263" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb55-264"><a href="#cb55-264" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of genes per bin"</span>) <span class="sc">+</span></span>
<span id="cb55-265"><a href="#cb55-265" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Pre-QC Genes/Bin'</span>) <span class="sc">+</span></span>
<span id="cb55-266"><a href="#cb55-266" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb55-267"><a href="#cb55-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-268"><a href="#cb55-268" aria-hidden="true" tabindex="-1"></a>dists_before <span class="ot">&lt;-</span> dist_counts_before <span class="sc">|</span> dist_features_before</span>
<span id="cb55-269"><a href="#cb55-269" aria-hidden="true" tabindex="-1"></a>dists_before</span>
<span id="cb55-270"><a href="#cb55-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-271"><a href="#cb55-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-272"><a href="#cb55-272" aria-hidden="true" tabindex="-1"></a>:::{.callout-tip}</span>
<span id="cb55-273"><a href="#cb55-273" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercise</span></span>
<span id="cb55-274"><a href="#cb55-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-275"><a href="#cb55-275" aria-hidden="true" tabindex="-1"></a>Using the distribution plots in the app below, what do you think would be good minimum thresholds for nGene and nUMI? </span>
<span id="cb55-276"><a href="#cb55-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-277"><a href="#cb55-277" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;&lt;</span><span class="kw">iframe</span><span class="ot"> src</span><span class="op">=</span><span class="st">"https://hcbc.connect.hms.harvard.edu/Spatial_threshold_question_2/?showcase=0"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800px"</span><span class="ot"> height</span><span class="op">=</span><span class="st">"410px"</span><span class="ot"> data-external</span><span class="op">=</span><span class="st">"1"</span><span class="dt">&gt;</span> <span class="dt">&lt;/</span><span class="kw">iframe</span><span class="dt">&gt;&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-278"><a href="#cb55-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-279"><a href="#cb55-279" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-280"><a href="#cb55-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-281"><a href="#cb55-281" aria-hidden="true" tabindex="-1"></a><span class="fu">### Post-Filtering</span></span>
<span id="cb55-282"><a href="#cb55-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-283"><a href="#cb55-283" aria-hidden="true" tabindex="-1"></a>We will apply very minimal filtering here, with nUMI &gt; 100 and nGene &gt; 100. It has been shown that low expression can be biologically meaningful for spatial context so we won't be as stringent as we normally are with scRNA-seq.</span>
<span id="cb55-284"><a href="#cb55-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-285"><a href="#cb55-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-288"><a href="#cb55-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-289"><a href="#cb55-289" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a filtered object</span></span>
<span id="cb55-290"><a href="#cb55-290" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(object, </span>
<span id="cb55-291"><a href="#cb55-291" aria-hidden="true" tabindex="-1"></a>                      <span class="at">subset =</span> (nCount_Spatial<span class="fl">.016</span>um <span class="sc">&gt;</span> <span class="dv">100</span>) <span class="sc">&amp;</span> </span>
<span id="cb55-292"><a href="#cb55-292" aria-hidden="true" tabindex="-1"></a>                        (nFeature_Spatial<span class="fl">.016</span>um <span class="sc">&gt;</span> <span class="dv">100</span>))</span>
<span id="cb55-293"><a href="#cb55-293" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-294"><a href="#cb55-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-295"><a href="#cb55-295" aria-hidden="true" tabindex="-1"></a>Now, we can create similar plots with filtered data. As expected, we see that the small left peak in the distribution has vanished, leaving the higher quality bins, which are the majority of the data. </span>
<span id="cb55-296"><a href="#cb55-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-299"><a href="#cb55-299" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-300"><a href="#cb55-300" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new metadata data frame </span></span>
<span id="cb55-301"><a href="#cb55-301" aria-hidden="true" tabindex="-1"></a>object_filt_meta <span class="ot">&lt;-</span> object_filt<span class="sc">@</span>meta.data</span>
<span id="cb55-302"><a href="#cb55-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-303"><a href="#cb55-303" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot nUMI</span></span>
<span id="cb55-304"><a href="#cb55-304" aria-hidden="true" tabindex="-1"></a>dist_counts_after <span class="ot">&lt;-</span> object_filt_meta <span class="sc">%&gt;%</span></span>
<span id="cb55-305"><a href="#cb55-305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nCount_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb55-306"><a href="#cb55-306" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb55-307"><a href="#cb55-307" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb55-308"><a href="#cb55-308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb55-309"><a href="#cb55-309" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb55-310"><a href="#cb55-310" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of UMIs per bin"</span>) <span class="sc">+</span></span>
<span id="cb55-311"><a href="#cb55-311" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'PostQC UMIs/Bin'</span>) <span class="sc">+</span></span>
<span id="cb55-312"><a href="#cb55-312" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb55-313"><a href="#cb55-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-314"><a href="#cb55-314" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot nGene</span></span>
<span id="cb55-315"><a href="#cb55-315" aria-hidden="true" tabindex="-1"></a>dist_features_after <span class="ot">&lt;-</span> object_filt_meta <span class="sc">%&gt;%</span></span>
<span id="cb55-316"><a href="#cb55-316" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nFeature_Spatial<span class="fl">.016</span>um)) <span class="sc">+</span></span>
<span id="cb55-317"><a href="#cb55-317" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb55-318"><a href="#cb55-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb55-319"><a href="#cb55-319" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb55-320"><a href="#cb55-320" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb55-321"><a href="#cb55-321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Number of genes per bin"</span>) <span class="sc">+</span></span>
<span id="cb55-322"><a href="#cb55-322" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'PostQC Genes/Bin'</span>) <span class="sc">+</span></span>
<span id="cb55-323"><a href="#cb55-323" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb55-324"><a href="#cb55-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-325"><a href="#cb55-325" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine plots side-by-side</span></span>
<span id="cb55-326"><a href="#cb55-326" aria-hidden="true" tabindex="-1"></a>dists_after <span class="ot">&lt;-</span> dist_counts_after <span class="sc">|</span> dist_features_after</span>
<span id="cb55-327"><a href="#cb55-327" aria-hidden="true" tabindex="-1"></a>dists_after</span>
<span id="cb55-328"><a href="#cb55-328" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-329"><a href="#cb55-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-330"><a href="#cb55-330" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing Counts Data</span></span>
<span id="cb55-331"><a href="#cb55-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-332"><a href="#cb55-332" aria-hidden="true" tabindex="-1"></a>We can visualize the number of UMIs and gene counts per bin, both as a distribution and layered on top of the tissue image. Let's start with a violin plot to look at the distribution of UMI counts and gene counts. The input is our post-filtered dataset.</span>
<span id="cb55-333"><a href="#cb55-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-336"><a href="#cb55-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-337"><a href="#cb55-337" aria-hidden="true" tabindex="-1"></a><span class="co"># Violin plot of UMI counts</span></span>
<span id="cb55-338"><a href="#cb55-338" aria-hidden="true" tabindex="-1"></a>vln_counts_after <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(object_filt, </span>
<span id="cb55-339"><a href="#cb55-339" aria-hidden="true" tabindex="-1"></a>                            <span class="at">features =</span> <span class="st">"nCount_Spatial.016um"</span>, </span>
<span id="cb55-340"><a href="#cb55-340" aria-hidden="true" tabindex="-1"></a>                            <span class="at">pt.size =</span> <span class="dv">0</span>, </span>
<span id="cb55-341"><a href="#cb55-341" aria-hidden="true" tabindex="-1"></a>                            <span class="at">group.by =</span> <span class="st">'orig.ident'</span>) <span class="sc">+</span> </span>
<span id="cb55-342"><a href="#cb55-342" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'nUMI'</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">''</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">15000</span>))</span>
<span id="cb55-343"><a href="#cb55-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-344"><a href="#cb55-344" aria-hidden="true" tabindex="-1"></a><span class="co"># Violin plot of gene counts</span></span>
<span id="cb55-345"><a href="#cb55-345" aria-hidden="true" tabindex="-1"></a>vln_features_after <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(object_filt, </span>
<span id="cb55-346"><a href="#cb55-346" aria-hidden="true" tabindex="-1"></a>                            <span class="at">features =</span> <span class="st">"nFeature_Spatial.016um"</span>, </span>
<span id="cb55-347"><a href="#cb55-347" aria-hidden="true" tabindex="-1"></a>                            <span class="at">pt.size =</span> <span class="dv">0</span>, </span>
<span id="cb55-348"><a href="#cb55-348" aria-hidden="true" tabindex="-1"></a>                            <span class="at">group.by =</span> <span class="st">'orig.ident'</span>) <span class="sc">+</span> </span>
<span id="cb55-349"><a href="#cb55-349" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'nGene'</span>) <span class="sc">+</span>  <span class="fu">xlab</span>(<span class="st">''</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">15000</span>))</span>
<span id="cb55-350"><a href="#cb55-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-351"><a href="#cb55-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-352"><a href="#cb55-352" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot both side by side</span></span>
<span id="cb55-353"><a href="#cb55-353" aria-hidden="true" tabindex="-1"></a>vln_counts_after <span class="sc">|</span> vln_features_after</span>
<span id="cb55-354"><a href="#cb55-354" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-355"><a href="#cb55-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-356"><a href="#cb55-356" aria-hidden="true" tabindex="-1"></a>We see that both distributions have a similar peak but the nUMI distribution has a much longer tail. This is expected, because while the small physical size of the bins means that most genes will be detected only once or twice, a minority of bins under very transcriptionally active cells may exhibit multiple transcripts of the same gene. </span>
<span id="cb55-357"><a href="#cb55-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-358"><a href="#cb55-358" aria-hidden="true" tabindex="-1"></a>Next, we can look at the same metrics and the distribution on the actual image itself. Note that many spots have very few counts, in part due to low cellular density or cell types with low complexity in certain tissue regions.</span>
<span id="cb55-359"><a href="#cb55-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-362"><a href="#cb55-362" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-363"><a href="#cb55-363" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing UMI count across the image</span></span>
<span id="cb55-364"><a href="#cb55-364" aria-hidden="true" tabindex="-1"></a>image_counts <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(object_filt, </span>
<span id="cb55-365"><a href="#cb55-365" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">feature =</span> <span class="st">'nCount_Spatial.016um'</span>, </span>
<span id="cb55-366"><a href="#cb55-366" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">pt.size.factor =</span> <span class="dv">8</span>)</span>
<span id="cb55-367"><a href="#cb55-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-368"><a href="#cb55-368" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing gene count across the image</span></span>
<span id="cb55-369"><a href="#cb55-369" aria-hidden="true" tabindex="-1"></a>image_features <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(object_filt, </span>
<span id="cb55-370"><a href="#cb55-370" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">features =</span> <span class="st">"nFeature_Spatial.016um"</span>, </span>
<span id="cb55-371"><a href="#cb55-371" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">pt.size.factor =</span> <span class="dv">8</span>) </span>
<span id="cb55-372"><a href="#cb55-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-373"><a href="#cb55-373" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the two side-by-side</span></span>
<span id="cb55-374"><a href="#cb55-374" aria-hidden="true" tabindex="-1"></a>image_counts <span class="sc">|</span> image_features</span>
<span id="cb55-375"><a href="#cb55-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-376"><a href="#cb55-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-377"><a href="#cb55-377" aria-hidden="true" tabindex="-1"></a><span class="fu">## Normalize Data</span></span>
<span id="cb55-378"><a href="#cb55-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-379"><a href="#cb55-379" aria-hidden="true" tabindex="-1"></a>Normalization is important in order to make expression counts comparable across genes and/or samples. We note that the best normalization methods for spatial data are still being developed and evaluated. In particular, <span class="co">[</span><span class="ot">Bhuva et. al</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03241-7)</span> tested a variety of normalization techniques and found that normalizing by the number of transcripts detected per bin negatively affects spatial domain identification because total detections per bin can represent real biology. We are cognizant of this, but as discussed earlier, it can be challenging to determine whether a bin has few detections because of a technical artifact or biological signal. In the absence of normalization, this lack of signal will strongly affect clustering regardless of whether it is technical or biological. For this reason, we apply a standard log-transformed library size normalization to our data. </span>
<span id="cb55-380"><a href="#cb55-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-383"><a href="#cb55-383" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-384"><a href="#cb55-384" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(object_filt, <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>)</span>
<span id="cb55-385"><a href="#cb55-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-386"><a href="#cb55-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-387"><a href="#cb55-387" aria-hidden="true" tabindex="-1"></a>After normalization, we can call our Seurat object with:</span>
<span id="cb55-388"><a href="#cb55-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-391"><a href="#cb55-391" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-392"><a href="#cb55-392" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-393"><a href="#cb55-393" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-394"><a href="#cb55-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-395"><a href="#cb55-395" aria-hidden="true" tabindex="-1"></a>And we can see that there is now a new "data" layer in the Seurat object. </span>
<span id="cb55-396"><a href="#cb55-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-397"><a href="#cb55-397" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-398"><a href="#cb55-398" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_with_normalized_data_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-399"><a href="#cb55-399" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-400"><a href="#cb55-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-401"><a href="#cb55-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-402"><a href="#cb55-402" aria-hidden="true" tabindex="-1"></a><span class="fu">## Unsupervised Clustering</span></span>
<span id="cb55-403"><a href="#cb55-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-404"><a href="#cb55-404" aria-hidden="true" tabindex="-1"></a>The authors of the Seurat package recommend the Seurat v5 sketch clustering workflow because it exhibits improved performance for large datasets, especially for identifying rare and spatially-restricted groups. Sketch-based analyses aim to "subsample" large datasets in a way that preserves rare populations. </span>
<span id="cb55-405"><a href="#cb55-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-406"><a href="#cb55-406" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-407"><a href="#cb55-407" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Standard_clustering.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800"</span><span class="dt">&gt;</span></span>
<span id="cb55-408"><a href="#cb55-408" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-409"><a href="#cb55-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-410"><a href="#cb55-410" aria-hidden="true" tabindex="-1"></a>We will start by defining a set of highly variable genes. _Note that this is being done on all bins in our object._ Using this list of genes will help us to quantify the variability and similarity between bins. </span>
<span id="cb55-411"><a href="#cb55-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-414"><a href="#cb55-414" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-415"><a href="#cb55-415" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(object_filt)</span>
<span id="cb55-416"><a href="#cb55-416" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-417"><a href="#cb55-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-418"><a href="#cb55-418" aria-hidden="true" tabindex="-1"></a>We can examine our Seurat object and see that ```FindVariableFeatures()``` has added 2,000 variable features.</span>
<span id="cb55-419"><a href="#cb55-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-422"><a href="#cb55-422" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-423"><a href="#cb55-423" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-424"><a href="#cb55-424" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-425"><a href="#cb55-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-426"><a href="#cb55-426" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-427"><a href="#cb55-427" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_variable_features_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-428"><a href="#cb55-428" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-429"><a href="#cb55-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-430"><a href="#cb55-430" aria-hidden="true" tabindex="-1"></a>Next, we select 10,000 cells and create a new sub-sampled "sketch" assay using the <span class="in">`SketchData()`</span> function. The function takes a normalized single-cell dataset containing a set of variable features and returns a Seurat object with a new assay (sketch), consisting of 10,000 bins selected based off a "leverage score" for each bin. The leverage score reflects the magnitude of its contribution to the gene-covariance matrix, and its importance to the overall dataset, with rare populations earning a higher leverage score. This means that our 10,000 cells selected for the sketch will oversample rare populations, retaining the biological complexity of the sample, while drastically compressing the dataset.</span>
<span id="cb55-431"><a href="#cb55-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-434"><a href="#cb55-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-435"><a href="#cb55-435" aria-hidden="true" tabindex="-1"></a><span class="co"># we select 10,000 cells and create a new 'sketch' assay</span></span>
<span id="cb55-436"><a href="#cb55-436" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">SketchData</span>(</span>
<span id="cb55-437"><a href="#cb55-437" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> object_filt,</span>
<span id="cb55-438"><a href="#cb55-438" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>,</span>
<span id="cb55-439"><a href="#cb55-439" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncells =</span> <span class="dv">10000</span>,</span>
<span id="cb55-440"><a href="#cb55-440" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"LeverageScore"</span>,</span>
<span id="cb55-441"><a href="#cb55-441" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span></span>
<span id="cb55-442"><a href="#cb55-442" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-443"><a href="#cb55-443" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-444"><a href="#cb55-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-445"><a href="#cb55-445" aria-hidden="true" tabindex="-1"></a>Now that we have the sketched data, we can call the Seurat object:</span>
<span id="cb55-446"><a href="#cb55-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-449"><a href="#cb55-449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-450"><a href="#cb55-450" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-451"><a href="#cb55-451" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-452"><a href="#cb55-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-453"><a href="#cb55-453" aria-hidden="true" tabindex="-1"></a>We will see that there are four major changes that have taken place:</span>
<span id="cb55-454"><a href="#cb55-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-455"><a href="#cb55-455" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The number of features in the second line has double, because we have added a new assay</span>
<span id="cb55-456"><a href="#cb55-456" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Accordingly, the number of assays has increased from one to two</span>
<span id="cb55-457"><a href="#cb55-457" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The active assay has change from <span class="in">`Spatial.016um`</span> to <span class="in">`sketch`</span></span>
<span id="cb55-458"><a href="#cb55-458" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>There is a new line listing additional assays that exist in the Seurat object</span>
<span id="cb55-459"><a href="#cb55-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-460"><a href="#cb55-460" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-461"><a href="#cb55-461" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_sketch_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-462"><a href="#cb55-462" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-463"><a href="#cb55-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-464"><a href="#cb55-464" aria-hidden="true" tabindex="-1"></a>We can also see that the leverage score has been added as a column to the metadata of our object.</span>
<span id="cb55-465"><a href="#cb55-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-468"><a href="#cb55-468" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-469"><a href="#cb55-469" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb55-470"><a href="#cb55-470" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(object_filt<span class="sc">@</span>meta.data)</span>
<span id="cb55-471"><a href="#cb55-471" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-472"><a href="#cb55-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-475"><a href="#cb55-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-476"><a href="#cb55-476" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb55-477"><a href="#cb55-477" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DT)</span>
<span id="cb55-478"><a href="#cb55-478" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(object_filt<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">30</span>))</span>
<span id="cb55-479"><a href="#cb55-479" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-480"><a href="#cb55-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-481"><a href="#cb55-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-482"><a href="#cb55-482" aria-hidden="true" tabindex="-1"></a>Next, we will peform a standard clustering workflow on our sketch of 10,000 cells:</span>
<span id="cb55-483"><a href="#cb55-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-484"><a href="#cb55-484" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`FindVariableFeatures`</span>: As before, this generates a list of highly variable genes, which may be slighly different for the sketched dataset than for the full dataset</span>
<span id="cb55-485"><a href="#cb55-485" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`ScaleData`</span>: Highly variable genes will be confounded with the most highly expressed genes, so we need to adjust for this</span>
<span id="cb55-486"><a href="#cb55-486" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`RunPCA`</span>: Perform a principal component analysis using our scaled data and variable genes. This will emphasize variation in gene expression as well as similarity across bins</span>
<span id="cb55-487"><a href="#cb55-487" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`FindNeighbors`</span>: Determine the Euclidean distance between bins in PCA space</span>
<span id="cb55-488"><a href="#cb55-488" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`FindClusters`</span>: Iteratively group bins together based on neighborhood distances. Higher resolution will yield more groups. </span>
<span id="cb55-489"><a href="#cb55-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-490"><a href="#cb55-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-493"><a href="#cb55-493" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-494"><a href="#cb55-494" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(object_filt)</span>
<span id="cb55-495"><a href="#cb55-495" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(object_filt)</span>
<span id="cb55-496"><a href="#cb55-496" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(object_filt, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb55-497"><a href="#cb55-497" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction.name =</span> <span class="st">"pca.sketch"</span>)</span>
<span id="cb55-498"><a href="#cb55-498" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(object_filt, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb55-499"><a href="#cb55-499" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction =</span> <span class="st">"pca.sketch"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb55-500"><a href="#cb55-500" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(object_filt, </span>
<span id="cb55-501"><a href="#cb55-501" aria-hidden="true" tabindex="-1"></a>                            <span class="at">cluster.name =</span> <span class="st">"seurat_cluster.sketched"</span>, </span>
<span id="cb55-502"><a href="#cb55-502" aria-hidden="true" tabindex="-1"></a>                            <span class="at">resolution =</span> .<span class="dv">65</span>)</span>
<span id="cb55-503"><a href="#cb55-503" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-504"><a href="#cb55-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-505"><a href="#cb55-505" aria-hidden="true" tabindex="-1"></a>Finally, let's create a UMAP using the principal components as input. UMAP is a method that aims to place cells with similar local neighborhoods in high-dimensional space together in low-dimensional space, which is useful for visualizing our newly calculated clusters. We observe good separation between groups annotated as separate clusters, which is sign that our clustering indeed represents various cell types. </span>
<span id="cb55-506"><a href="#cb55-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-509"><a href="#cb55-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-510"><a href="#cb55-510" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(object_filt, <span class="at">reduction =</span> <span class="st">"pca.sketch"</span>, </span>
<span id="cb55-511"><a href="#cb55-511" aria-hidden="true" tabindex="-1"></a>                       <span class="at">reduction.name =</span> <span class="st">"umap.sketch"</span>, <span class="at">return.model =</span> T, </span>
<span id="cb55-512"><a href="#cb55-512" aria-hidden="true" tabindex="-1"></a>                       <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb55-513"><a href="#cb55-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-514"><a href="#cb55-514" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP</span></span>
<span id="cb55-515"><a href="#cb55-515" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(object_filt, <span class="at">reduction =</span> <span class="st">"umap.sketch"</span>, <span class="at">label =</span> T, </span>
<span id="cb55-516"><a href="#cb55-516" aria-hidden="true" tabindex="-1"></a>        <span class="at">cols =</span> <span class="st">'polychrome'</span>) <span class="sc">+</span> </span>
<span id="cb55-517"><a href="#cb55-517" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Sketched clustering"</span>) <span class="sc">+</span> </span>
<span id="cb55-518"><a href="#cb55-518" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb55-519"><a href="#cb55-519" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-520"><a href="#cb55-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-521"><a href="#cb55-521" aria-hidden="true" tabindex="-1"></a>We can also examine our object after these manipulations and note the addition of the scale.data layer as well as the sketch PCA and UMAP dimensional reductions. </span>
<span id="cb55-522"><a href="#cb55-522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-525"><a href="#cb55-525" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-526"><a href="#cb55-526" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-527"><a href="#cb55-527" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-528"><a href="#cb55-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-529"><a href="#cb55-529" aria-hidden="true" tabindex="-1"></a>Should return:</span>
<span id="cb55-530"><a href="#cb55-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-531"><a href="#cb55-531" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-532"><a href="#cb55-532" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_scale_data_PCA_UMAP_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-533"><a href="#cb55-533" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-534"><a href="#cb55-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-535"><a href="#cb55-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-536"><a href="#cb55-536" aria-hidden="true" tabindex="-1"></a><span class="fu">## Project cluster labels back to the full dataset</span></span>
<span id="cb55-537"><a href="#cb55-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-538"><a href="#cb55-538" aria-hidden="true" tabindex="-1"></a>Now that we have our clusters and dimensional reductions from our sketched dataset, we need to extend these to the full dataset.  The <span class="in">`ProjectData`</span> function projects all the bins in the dataset (the <span class="in">`Spatial.016um`</span> assay) onto the <span class="in">`sketch`</span> assay. </span>
<span id="cb55-539"><a href="#cb55-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-542"><a href="#cb55-542" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-543"><a href="#cb55-543" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">ProjectData</span>(</span>
<span id="cb55-544"><a href="#cb55-544" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> object_filt,</span>
<span id="cb55-545"><a href="#cb55-545" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"Spatial.016um"</span>,</span>
<span id="cb55-546"><a href="#cb55-546" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.reduction =</span> <span class="st">"full.pca.sketch"</span>,</span>
<span id="cb55-547"><a href="#cb55-547" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span>,</span>
<span id="cb55-548"><a href="#cb55-548" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.reduction =</span> <span class="st">"pca.sketch"</span>,</span>
<span id="cb55-549"><a href="#cb55-549" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.model =</span> <span class="st">"umap.sketch"</span>,</span>
<span id="cb55-550"><a href="#cb55-550" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb55-551"><a href="#cb55-551" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">seurat_cluster.projected =</span> <span class="st">"seurat_cluster.sketched"</span>)</span>
<span id="cb55-552"><a href="#cb55-552" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-553"><a href="#cb55-553" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-554"><a href="#cb55-554" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-555"><a href="#cb55-555" aria-hidden="true" tabindex="-1"></a>Using the sketch PCA and UMAP, the <span class="in">`ProjectData`</span> function returns a Seurat object that includes:</span>
<span id="cb55-556"><a href="#cb55-556" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>**Dimensional reduction (PCA)** - The <span class="in">`full.pca.sketch`</span> dimensional reduction extends the PCA reduction on the sketched cells to all bins in the dataset</span>
<span id="cb55-557"><a href="#cb55-557" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>**Dimensional reduction (UMAP)** - The <span class="in">`full.umap.sketch`</span> dimensional reduction extends the UMAP reduction on the sketched cells to all bins in the dataset</span>
<span id="cb55-558"><a href="#cb55-558" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>**Cluster labels** - The <span class="in">`seurat_cluster.projected`</span> column in the object metadata now labels all cells in the dataset with one of the cluster labels derived from the sketched cells</span>
<span id="cb55-559"><a href="#cb55-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-560"><a href="#cb55-560" aria-hidden="true" tabindex="-1"></a>We can now see the additional full-dataset reductions in the object.</span>
<span id="cb55-561"><a href="#cb55-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-564"><a href="#cb55-564" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-565"><a href="#cb55-565" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-566"><a href="#cb55-566" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-567"><a href="#cb55-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-568"><a href="#cb55-568" aria-hidden="true" tabindex="-1"></a>Should return:</span>
<span id="cb55-569"><a href="#cb55-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-570"><a href="#cb55-570" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-571"><a href="#cb55-571" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_projected_data_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-572"><a href="#cb55-572" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-573"><a href="#cb55-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-574"><a href="#cb55-574" aria-hidden="true" tabindex="-1"></a>Note that a score for the projection of each bin will be saved as a column in the metadata. Actually opening up the metadata again gives the opportunity to look at the <span class="in">`seurat_cluster.sketched`</span> column and see many NA values, because it was only calculated for 10,000 bins. The <span class="in">`seurat_cluster.projected`</span> shows values for every bin.</span>
<span id="cb55-575"><a href="#cb55-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-578"><a href="#cb55-578" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-579"><a href="#cb55-579" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb55-580"><a href="#cb55-580" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(object_filt<span class="sc">@</span>meta.data)</span>
<span id="cb55-581"><a href="#cb55-581" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-582"><a href="#cb55-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-585"><a href="#cb55-585" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-586"><a href="#cb55-586" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb55-587"><a href="#cb55-587" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DT)</span>
<span id="cb55-588"><a href="#cb55-588" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(object_filt<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">30</span>))</span>
<span id="cb55-589"><a href="#cb55-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-590"><a href="#cb55-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-591"><a href="#cb55-591" aria-hidden="true" tabindex="-1"></a>Should return:</span>
<span id="cb55-592"><a href="#cb55-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-593"><a href="#cb55-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-594"><a href="#cb55-594" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing the projected clusters on UMAP</span></span>
<span id="cb55-595"><a href="#cb55-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-596"><a href="#cb55-596" aria-hidden="true" tabindex="-1"></a>We can now visualize our clusters from the projected assignments. The UMAP plot now contains more points, which is expected because we are now visualizing the full dataset rather than our 10,000 bin sketch. Nonetheless, we can see that the full dataset is still well-representated by the projected dimensional reduction and clustering. </span>
<span id="cb55-597"><a href="#cb55-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-600"><a href="#cb55-600" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-601"><a href="#cb55-601" aria-hidden="true" tabindex="-1"></a><span class="co"># switch to full dataset assay</span></span>
<span id="cb55-602"><a href="#cb55-602" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(object_filt) <span class="ot">&lt;-</span> <span class="st">"Spatial.016um"</span></span>
<span id="cb55-603"><a href="#cb55-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-604"><a href="#cb55-604" aria-hidden="true" tabindex="-1"></a><span class="co"># Change the idents to the projected cluster assignments</span></span>
<span id="cb55-605"><a href="#cb55-605" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(object_filt) <span class="ot">&lt;-</span> <span class="st">"seurat_cluster.projected"</span></span>
<span id="cb55-606"><a href="#cb55-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-607"><a href="#cb55-607" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb55-608"><a href="#cb55-608" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(object_filt, <span class="at">reduction =</span> <span class="st">"full.umap.sketch"</span>, <span class="at">label =</span> T, <span class="at">raster =</span> F, </span>
<span id="cb55-609"><a href="#cb55-609" aria-hidden="true" tabindex="-1"></a>              <span class="at">cols =</span> <span class="st">'polychrome'</span>) <span class="sc">+</span></span>
<span id="cb55-610"><a href="#cb55-610" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Projected clustering"</span>) <span class="sc">+</span> </span>
<span id="cb55-611"><a href="#cb55-611" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb55-612"><a href="#cb55-612" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-613"><a href="#cb55-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-614"><a href="#cb55-614" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing projected clusters on the image</span></span>
<span id="cb55-615"><a href="#cb55-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-616"><a href="#cb55-616" aria-hidden="true" tabindex="-1"></a>In order to see the clusters superimposed on our image we can use the <span class="in">`SpatialDimPlot()`</span> function. We will also set the color palette and convert the cluster assignments to a factor so they are ordered numerically rather than lexicographically in the figure.</span>
<span id="cb55-617"><a href="#cb55-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-620"><a href="#cb55-620" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-621"><a href="#cb55-621" aria-hidden="true" tabindex="-1"></a><span class="co"># Arrange so clusters get listed in numerical order</span></span>
<span id="cb55-622"><a href="#cb55-622" aria-hidden="true" tabindex="-1"></a>object_filt<span class="sc">$</span>seurat_cluster.projected <span class="ot">&lt;-</span> object_filt<span class="sc">$</span>seurat_cluster.projected <span class="sc">%&gt;%</span> </span>
<span id="cb55-623"><a href="#cb55-623" aria-hidden="true" tabindex="-1"></a>  as.numeric <span class="sc">%&gt;%</span> <span class="fu">as.factor</span>()</span>
<span id="cb55-624"><a href="#cb55-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-625"><a href="#cb55-625" aria-hidden="true" tabindex="-1"></a><span class="co"># Set color palette</span></span>
<span id="cb55-626"><a href="#cb55-626" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected)),</span>
<span id="cb55-627"><a href="#cb55-627" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb55-628"><a href="#cb55-628" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected))</span>
<span id="cb55-629"><a href="#cb55-629" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(object_filt, </span>
<span id="cb55-630"><a href="#cb55-630" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">group.by =</span> <span class="st">'seurat_cluster.projected'</span>, </span>
<span id="cb55-631"><a href="#cb55-631" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">pt.size.factor =</span> <span class="dv">8</span>, <span class="at">cols =</span> color_pal) <span class="sc">+</span></span>
<span id="cb55-632"><a href="#cb55-632" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">ncol=</span><span class="dv">2</span>))</span>
<span id="cb55-633"><a href="#cb55-633" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-634"><a href="#cb55-634" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters</span>
<span id="cb55-635"><a href="#cb55-635" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-636"><a href="#cb55-636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-637"><a href="#cb55-637" aria-hidden="true" tabindex="-1"></a><span class="fu">## Spatially-informed Clustering</span></span>
<span id="cb55-638"><a href="#cb55-638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-639"><a href="#cb55-639" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">BANKSY</span><span class="co">](https://www.nature.com/articles/s41588-024-01664-3)</span> is another method for performing clustering. Unlike Seurat, BANKSY takes into account not only an individual bin’s expression pattern but also the mean and the gradient of gene expression levels in a bin’s broader neighborhood. This makes it valuable for identifying and defining spatial regions of interest.</span>
<span id="cb55-640"><a href="#cb55-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-641"><a href="#cb55-641" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`RunBanksy`</span> function to create a new "BANKSY" assay based on a default of the 4,000 most highly variable features, which can be used for dimensionality reduction and clustering. Two parameters of importance are:</span>
<span id="cb55-642"><a href="#cb55-642" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`k_geom`</span> - Number of bins to consider for the local neighborhood. Larger values will yield larger domains.</span>
<span id="cb55-643"><a href="#cb55-643" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`lambda`</span> - Influence of the neighborhood. Larger values yield more spatially coherent domains. The authors recommend using 0.8 to identify broader spatial domains. </span>
<span id="cb55-644"><a href="#cb55-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-647"><a href="#cb55-647" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-648"><a href="#cb55-648" aria-hidden="true" tabindex="-1"></a><span class="co"># Run Banksy</span></span>
<span id="cb55-649"><a href="#cb55-649" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunBanksy</span>(object_filt, <span class="at">lambda =</span> <span class="fl">0.8</span>, <span class="at">verbose =</span> T,</span>
<span id="cb55-650"><a href="#cb55-650" aria-hidden="true" tabindex="-1"></a>                         <span class="at">assay =</span> <span class="st">'Spatial.016um'</span>, <span class="at">slot =</span> <span class="st">'data'</span>, <span class="at">k_geom =</span> <span class="dv">50</span>)</span>
<span id="cb55-651"><a href="#cb55-651" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-652"><a href="#cb55-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-653"><a href="#cb55-653" aria-hidden="true" tabindex="-1"></a>We can see the new BANKSY assay in our object by calling:</span>
<span id="cb55-654"><a href="#cb55-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-657"><a href="#cb55-657" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-658"><a href="#cb55-658" aria-hidden="true" tabindex="-1"></a>object_filt</span>
<span id="cb55-659"><a href="#cb55-659" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-660"><a href="#cb55-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-661"><a href="#cb55-661" aria-hidden="true" tabindex="-1"></a>Which should return:</span>
<span id="cb55-662"><a href="#cb55-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-663"><a href="#cb55-663" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-664"><a href="#cb55-664" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Seurat_object_BANKSY_labelled.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"700"</span><span class="dt">&gt;</span></span>
<span id="cb55-665"><a href="#cb55-665" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-666"><a href="#cb55-666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-667"><a href="#cb55-667" aria-hidden="true" tabindex="-1"></a>We perform a simplified clustering workflow on the BANKSY assay.</span>
<span id="cb55-668"><a href="#cb55-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-671"><a href="#cb55-671" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-672"><a href="#cb55-672" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(object_filt, <span class="at">assay =</span> <span class="st">"BANKSY"</span>, </span>
<span id="cb55-673"><a href="#cb55-673" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction.name =</span> <span class="st">"pca.banksy"</span>, </span>
<span id="cb55-674"><a href="#cb55-674" aria-hidden="true" tabindex="-1"></a>                      <span class="at">features =</span> <span class="fu">rownames</span>(object_filt), <span class="at">npcs =</span> <span class="dv">30</span>)</span>
<span id="cb55-675"><a href="#cb55-675" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(object_filt, <span class="at">reduction =</span> <span class="st">"pca.banksy"</span>, </span>
<span id="cb55-676"><a href="#cb55-676" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb55-677"><a href="#cb55-677" aria-hidden="true" tabindex="-1"></a>object_filt <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(object_filt, <span class="at">cluster.name =</span> <span class="st">"banksy_cluster"</span>,</span>
<span id="cb55-678"><a href="#cb55-678" aria-hidden="true" tabindex="-1"></a>                            <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb55-679"><a href="#cb55-679" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-680"><a href="#cb55-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-681"><a href="#cb55-681" aria-hidden="true" tabindex="-1"></a>Let's visualize the BANKSY clusters alongside the Seurat clusters for a side-by-side comparison:</span>
<span id="cb55-682"><a href="#cb55-682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-685"><a href="#cb55-685" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-686"><a href="#cb55-686" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>banksy_cluster)),</span>
<span id="cb55-687"><a href="#cb55-687" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb55-688"><a href="#cb55-688" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>banksy_cluster))</span>
<span id="cb55-689"><a href="#cb55-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-690"><a href="#cb55-690" aria-hidden="true" tabindex="-1"></a>image_banksy_clusters <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(object_filt, <span class="at">group.by =</span> <span class="st">"banksy_cluster"</span>,</span>
<span id="cb55-691"><a href="#cb55-691" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">pt.size.factor =</span> <span class="dv">7</span>, <span class="at">cols =</span> color_pal)</span>
<span id="cb55-692"><a href="#cb55-692" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-693"><a href="#cb55-693" aria-hidden="true" tabindex="-1"></a>image_seurat_clusters <span class="sc">|</span> image_banksy_clusters</span>
<span id="cb55-694"><a href="#cb55-694" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-695"><a href="#cb55-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-696"><a href="#cb55-696" aria-hidden="true" tabindex="-1"></a>We can see that, as expected, the BANKSY clusters are more spatially-restricted, or more compact, than the Seurat clusters. We also see that the BANKSY clusters are less noisy than the Seurat clusters, likely because of the smoothing effect of considering a cell's spatial neighborhood when assigning a cluster label. </span>
<span id="cb55-697"><a href="#cb55-697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-698"><a href="#cb55-698" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse=true}</span>
<span id="cb55-699"><a href="#cb55-699" aria-hidden="true" tabindex="-1"></a><span class="fu"># Click here to see BANKSY using a lambda value of 0.2</span></span>
<span id="cb55-700"><a href="#cb55-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-701"><a href="#cb55-701" aria-hidden="true" tabindex="-1"></a>If we had run BANKSY with <span class="in">`lambda = 0.2`</span>, as recommended for cell type clustering instead of <span class="in">`lambda = 0.8`</span> for spatial domain clustering, the resultant clusters would be less spatially restricted (in other words less compact and more distributed throughout the image) and more similar to our Seurat clustering. Below is a figure using <span class="in">`lamba=0.2`</span> in BANKSY rather than <span class="in">`lamba=0.8`</span>.</span>
<span id="cb55-702"><a href="#cb55-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-703"><a href="#cb55-703" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-704"><a href="#cb55-704" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/banksy_clustering_lambda_0.2.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"450"</span><span class="dt">&gt;</span></span>
<span id="cb55-705"><a href="#cb55-705" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-706"><a href="#cb55-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-707"><a href="#cb55-707" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-708"><a href="#cb55-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-709"><a href="#cb55-709" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell Type Annotation</span></span>
<span id="cb55-710"><a href="#cb55-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-711"><a href="#cb55-711" aria-hidden="true" tabindex="-1"></a>Perhaps we are particularly interested in understanding the organization of cell types in the cortical region of the brain. </span>
<span id="cb55-712"><a href="#cb55-712" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-713"><a href="#cb55-713" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">p</span><span class="ot"> align</span><span class="op">=</span><span class="st">"center"</span><span class="dt">&gt;</span></span>
<span id="cb55-714"><a href="#cb55-714" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">img</span><span class="ot"> src</span><span class="op">=</span><span class="st">"../img/Cell_type_annotations.png"</span><span class="ot"> width</span><span class="op">=</span><span class="st">"800"</span><span class="dt">&gt;</span></span>
<span id="cb55-715"><a href="#cb55-715" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">p</span><span class="dt">&gt;</span></span>
<span id="cb55-716"><a href="#cb55-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-717"><a href="#cb55-717" aria-hidden="true" tabindex="-1"></a>We first subset our Seurat object to this region of interest. There are multiple ways of subsetting a Seurat object to a region of interest, but here we have identified a handful of cluster numbers that appear almost exclusively in the cortical region, and we will subset the object to only include cells that are assigned these cluster numbers. </span>
<span id="cb55-718"><a href="#cb55-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-721"><a href="#cb55-721" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-722"><a href="#cb55-722" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">subset</span>(object_filt, seurat_cluster.projected <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">18</span>, <span class="dv">19</span>, <span class="dv">7</span>, <span class="dv">2</span>, <span class="dv">4</span>))</span>
<span id="cb55-723"><a href="#cb55-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-724"><a href="#cb55-724" aria-hidden="true" tabindex="-1"></a>color_pal <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">DiscretePalette</span>(<span class="at">n =</span> <span class="fu">length</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected)),</span>
<span id="cb55-725"><a href="#cb55-725" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">palette =</span> <span class="st">"polychrome"</span>)</span>
<span id="cb55-726"><a href="#cb55-726" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color_pal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(object_filt<span class="sc">$</span>seurat_cluster.projected))</span>
<span id="cb55-727"><a href="#cb55-727" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(cortex, <span class="at">group.by =</span> <span class="st">'seurat_cluster.projected'</span>, </span>
<span id="cb55-728"><a href="#cb55-728" aria-hidden="true" tabindex="-1"></a>               <span class="at">pt.size.factor =</span> <span class="dv">8</span>, <span class="at">cols =</span> color_pal)</span>
<span id="cb55-729"><a href="#cb55-729" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-730"><a href="#cb55-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-731"><a href="#cb55-731" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb55-732"><a href="#cb55-732" aria-hidden="true" tabindex="-1"></a>Note: Your colors may be different than the ones in the above figure.</span>
<span id="cb55-733"><a href="#cb55-733" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb55-734"><a href="#cb55-734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-735"><a href="#cb55-735" aria-hidden="true" tabindex="-1"></a>To perform accurate annotation of cell types, we must also take into consideration that our 16µm x 16µm bins may contain one or more cells each. The method <span class="co">[</span><span class="ot">Robust Cell Type Deconvolution</span><span class="co">](https://www.nature.com/articles/s41587-021-00830-w)</span> (RCTD) has been shown to accurately annotate spatial data from a variety of technologies while taking into consideration that a single bin may exhibit multiple cell type profiles.</span>
<span id="cb55-736"><a href="#cb55-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-737"><a href="#cb55-737" aria-hidden="true" tabindex="-1"></a>RCTD takes a cell-type-annotated scRNA-seq dataset as a reference and a spatial dataset as a query. For our reference, we will use a subsampled version of the mouse scRNA-seq dataset from the Allen Brain Atlas. We will use our cortex Seurat object as the spatial query. As an overview, the process is as follows:</span>
<span id="cb55-738"><a href="#cb55-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-739"><a href="#cb55-739" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Sketch and process the spatial query dataset</span>
<span id="cb55-740"><a href="#cb55-740" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Load and format the scRNA-seq reference dataset</span>
<span id="cb55-741"><a href="#cb55-741" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Apply RCTD to deconvolute the "sketched" cortical cells and annotate them</span>
<span id="cb55-742"><a href="#cb55-742" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Project these annotations onto the full cortical dataset.</span>
<span id="cb55-743"><a href="#cb55-743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-744"><a href="#cb55-744" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1) Sketch and process the spatial query dataset</span></span>
<span id="cb55-745"><a href="#cb55-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-748"><a href="#cb55-748" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-749"><a href="#cb55-749" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">'Spatial.016um'</span></span>
<span id="cb55-750"><a href="#cb55-750" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cortex)</span>
<span id="cb55-751"><a href="#cb55-751" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">SketchData</span>(</span>
<span id="cb55-752"><a href="#cb55-752" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb55-753"><a href="#cb55-753" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncells =</span> <span class="dv">3000</span>,</span>
<span id="cb55-754"><a href="#cb55-754" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"LeverageScore"</span>,</span>
<span id="cb55-755"><a href="#cb55-755" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span></span>
<span id="cb55-756"><a href="#cb55-756" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-757"><a href="#cb55-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-758"><a href="#cb55-758" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"sketch"</span></span>
<span id="cb55-759"><a href="#cb55-759" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cortex)</span>
<span id="cb55-760"><a href="#cb55-760" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cortex, <span class="at">assay =</span> <span class="st">"sketch"</span>, </span>
<span id="cb55-761"><a href="#cb55-761" aria-hidden="true" tabindex="-1"></a>                 <span class="at">reduction.name =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">verbose =</span> T)</span>
<span id="cb55-762"><a href="#cb55-762" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb55-763"><a href="#cb55-763" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, </span>
<span id="cb55-764"><a href="#cb55-764" aria-hidden="true" tabindex="-1"></a>                  <span class="at">reduction.name =</span> <span class="st">"umap.cortex.sketch"</span>, <span class="at">return.model =</span> T, </span>
<span id="cb55-765"><a href="#cb55-765" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, <span class="at">verbose =</span> T)</span>
<span id="cb55-766"><a href="#cb55-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-767"><a href="#cb55-767" aria-hidden="true" tabindex="-1"></a><span class="co"># create the RCTD query object</span></span>
<span id="cb55-768"><a href="#cb55-768" aria-hidden="true" tabindex="-1"></a>counts_hd <span class="ot">&lt;-</span> cortex[[<span class="st">"sketch"</span>]]<span class="sc">$</span>counts</span>
<span id="cb55-769"><a href="#cb55-769" aria-hidden="true" tabindex="-1"></a>cortex_cells_hd <span class="ot">&lt;-</span> <span class="fu">colnames</span>(cortex[[<span class="st">"sketch"</span>]])</span>
<span id="cb55-770"><a href="#cb55-770" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(cortex)[cortex_cells_hd, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb55-771"><a href="#cb55-771" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">SpatialRNA</span>(coords, counts_hd, <span class="fu">colSums</span>(counts_hd))</span>
<span id="cb55-772"><a href="#cb55-772" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-773"><a href="#cb55-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-774"><a href="#cb55-774" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2) Load and format the reference dataset</span></span>
<span id="cb55-775"><a href="#cb55-775" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-778"><a href="#cb55-778" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-779"><a href="#cb55-779" aria-hidden="true" tabindex="-1"></a><span class="fu">mem.maxVSize</span>(<span class="dv">15000</span>)</span>
<span id="cb55-780"><a href="#cb55-780" aria-hidden="true" tabindex="-1"></a>ref_subset <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"data_processed/allen_scRNAseq_ref_subset.qs"</span>)</span>
<span id="cb55-781"><a href="#cb55-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-782"><a href="#cb55-782" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ref_subset) <span class="ot">&lt;-</span> <span class="st">"subclass_label"</span></span>
<span id="cb55-783"><a href="#cb55-783" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> ref_subset[[<span class="st">"RNA"</span>]]<span class="sc">$</span>counts</span>
<span id="cb55-784"><a href="#cb55-784" aria-hidden="true" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(ref_subset<span class="sc">$</span>subclass_label)</span>
<span id="cb55-785"><a href="#cb55-785" aria-hidden="true" tabindex="-1"></a>nUMI <span class="ot">&lt;-</span> ref_subset<span class="sc">$</span>nCount_RNA</span>
<span id="cb55-786"><a href="#cb55-786" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(cluster) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"/"</span>, <span class="st">"-"</span>, <span class="fu">levels</span>(cluster))</span>
<span id="cb55-787"><a href="#cb55-787" aria-hidden="true" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">droplevels</span>(cluster)</span>
<span id="cb55-788"><a href="#cb55-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-789"><a href="#cb55-789" aria-hidden="true" tabindex="-1"></a><span class="co"># create the RCTD reference object</span></span>
<span id="cb55-790"><a href="#cb55-790" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">Reference</span>(counts, cluster, nUMI)</span>
<span id="cb55-791"><a href="#cb55-791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-792"><a href="#cb55-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-793"><a href="#cb55-793" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3) Apply RCTD to deconvolute the "sketched" cortical cells and annotate them</span></span>
<span id="cb55-794"><a href="#cb55-794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-795"><a href="#cb55-795" aria-hidden="true" tabindex="-1"></a>Note that ```run.RCTD``` takes 10-15 minutes to complete on a laptop using 6 cores</span>
<span id="cb55-796"><a href="#cb55-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-799"><a href="#cb55-799" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-800"><a href="#cb55-800" aria-hidden="true" tabindex="-1"></a><span class="co"># run RCTD</span></span>
<span id="cb55-801"><a href="#cb55-801" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">create.RCTD</span>(query, reference, <span class="at">max_cores =</span> <span class="dv">6</span>)</span>
<span id="cb55-802"><a href="#cb55-802" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">run.RCTD</span>(RCTD, <span class="at">doublet_mode =</span> <span class="st">"doublet"</span>) <span class="co"># this command takes ~15 mins to run</span></span>
<span id="cb55-803"><a href="#cb55-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-804"><a href="#cb55-804" aria-hidden="true" tabindex="-1"></a><span class="co"># add results back to Seurat object</span></span>
<span id="cb55-805"><a href="#cb55-805" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(cortex, <span class="at">metadata =</span> RCTD<span class="sc">@</span>results<span class="sc">$</span>results_df)</span>
<span id="cb55-806"><a href="#cb55-806" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-807"><a href="#cb55-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-808"><a href="#cb55-808" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4) Project RCTD labels onto all cortical cells</span></span>
<span id="cb55-809"><a href="#cb55-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-812"><a href="#cb55-812" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-813"><a href="#cb55-813" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type <span class="ot">&lt;-</span> <span class="fu">as.character</span>(cortex<span class="sc">$</span>first_type)</span>
<span id="cb55-814"><a href="#cb55-814" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type[<span class="fu">is.na</span>(cortex<span class="sc">$</span>first_type)] <span class="ot">&lt;-</span> <span class="st">"Unknown"</span></span>
<span id="cb55-815"><a href="#cb55-815" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ProjectData</span>(</span>
<span id="cb55-816"><a href="#cb55-816" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb55-817"><a href="#cb55-817" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"Spatial.016um"</span>,</span>
<span id="cb55-818"><a href="#cb55-818" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.reduction =</span> <span class="st">"pca.cortex"</span>,</span>
<span id="cb55-819"><a href="#cb55-819" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span>,</span>
<span id="cb55-820"><a href="#cb55-820" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.reduction =</span> <span class="st">"pca.cortex.sketch"</span>,</span>
<span id="cb55-821"><a href="#cb55-821" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.model =</span> <span class="st">"umap.cortex.sketch"</span>,</span>
<span id="cb55-822"><a href="#cb55-822" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb55-823"><a href="#cb55-823" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">full_first_type =</span> <span class="st">"first_type"</span>)</span>
<span id="cb55-824"><a href="#cb55-824" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-825"><a href="#cb55-825" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-826"><a href="#cb55-826" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-827"><a href="#cb55-827" aria-hidden="true" tabindex="-1"></a>We can see that the excitatory neurons are located in layers at varying cortical depths, as expected.</span>
<span id="cb55-828"><a href="#cb55-828" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-831"><a href="#cb55-831" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-832"><a href="#cb55-832" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"full_first_type"</span></span>
<span id="cb55-833"><a href="#cb55-833" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">CellsByIdentities</span>(cortex)</span>
<span id="cb55-834"><a href="#cb55-834" aria-hidden="true" tabindex="-1"></a><span class="co"># Layered (starts with L), excitatory neurons in the cortex</span></span>
<span id="cb55-835"><a href="#cb55-835" aria-hidden="true" tabindex="-1"></a>excitatory_names <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">grep</span>(<span class="st">"^L.* CTX"</span>, <span class="fu">names</span>(cells), <span class="at">value =</span> <span class="cn">TRUE</span>))</span>
<span id="cb55-836"><a href="#cb55-836" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(cortex, <span class="at">cells.highlight =</span> cells[excitatory_names], </span>
<span id="cb55-837"><a href="#cb55-837" aria-hidden="true" tabindex="-1"></a>               <span class="at">cols.highlight =</span> <span class="fu">c</span>(<span class="st">"#FFFF00"</span>, <span class="st">"grey50"</span>), <span class="at">facet.highlight =</span> T, </span>
<span id="cb55-838"><a href="#cb55-838" aria-hidden="true" tabindex="-1"></a>               <span class="at">combine =</span> T, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">pt.size.factor =</span> <span class="dv">8</span>)</span>
<span id="cb55-839"><a href="#cb55-839" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-840"><a href="#cb55-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-841"><a href="#cb55-841" aria-hidden="true" tabindex="-1"></a><span class="fu"># Save R Session</span></span>
<span id="cb55-842"><a href="#cb55-842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-843"><a href="#cb55-843" aria-hidden="true" tabindex="-1"></a>At the end of any analysis, it is important to save your session information! This will let future users know which versions of packages are necessary in order to reproduce the same results.</span>
<span id="cb55-844"><a href="#cb55-844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-845"><a href="#cb55-845" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse=true}</span>
<span id="cb55-846"><a href="#cb55-846" aria-hidden="true" tabindex="-1"></a><span class="fu"># sessionInfo</span></span>
<span id="cb55-849"><a href="#cb55-849" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb55-850"><a href="#cb55-850" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb55-851"><a href="#cb55-851" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb55-852"><a href="#cb55-852" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>